---
title: "RMS biobank Riboseq reads stats"
output: html_notebook
date: 13-07-2022
---

If MultiQC stats have not been generated yet, run the following on an interactive
compute note on the HPC:

```{bash}
container_dir="/hpc/local/Rocky8/pmc_vanheesch/singularity_images"
name="RMS_all"
wd="/hpc/pmc_vanheesch/projects/jvandinter/rms_analysis/02_riboseq"

apptainer exec -B "/hpc:/hpc" --env "LC_ALL=C.UTF-8" \
 ${container_dir}/multiqc-1.11.sif multiqc \
 "${wd}/analysis" \
 --outdir "${wd}/analysis/multiqc" \
 --filename "${name}_multiqc.html"
```

```{r}
suppressPackageStartupMessages({
  library(stringr)
  library(ggbeeswarm)
  library(ggsci)
  library(tidyverse)
  library(ggpubr)
})

```

```{r parameters}

wd = "/hpc/pmc_vanheesch/projects/jvandinter/rms_analysis/02_riboseq/analysis"

savedir = "/hpc/pmc_vanheesch/projects/jvandinter/rms_analysis/02_riboseq/results/QC"

multiqc_name = "RMS_all"

colors <- ggsci::pal_jco("default", alpha = 0.7)(9)[c(1:6,8)]

```

# Introduction

In this notebook, I check the RiboseQC output of the RMS biobank samples (without custom transcriptome). In addition, I used the number of CDS reads in each sample to calculate the pooling ratios; this might give a more accurate pooling ratio than simply taking the number of 'passed' reads in each sample.

## Trimming

```{r cutadapt}

trim_stats <- read.delim(paste(wd,
                               "multiqc",
                               paste0(multiqc_name,"_multiqc_data"),
                               "multiqc_cutadapt.txt", sep ="/")) %>%
  dplyr::group_by(Sample) %>%
  dplyr::summarize(r_written = sum(r_written), 
            bp_processed = sum(bp_processed),
            quality_trimmed = sum(quality_trimmed),
            bp_written = sum(bp_written))

trim_plot <- trim_stats %>%
  dplyr::select(c("Sample","r_written","bp_processed","quality_trimmed","bp_written")) %>%
  tidyr::pivot_longer(cols = c(3,4)) %>%
  dplyr::mutate(Sample = factor(Sample, levels = trim_stats[order(trim_stats$r_written),]$Sample)) %>%
  dplyr::mutate(Sample = gsub("-.*","",Sample))

```

```{r cutadapt plot}
ggplot(data = trim_plot, aes(x = Sample, y = value, fill = name)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = rev(ggsci::pal_jco("default")(2)),
                    labels = c("BP processed", "BP trimmed")) +
  theme_classic() +
  labs(
       x = "Samples", 
       y = "Cumulative basepairs", 
       fill = "Read type") +
  theme(axis.text.x = element_text(angle = 70, hjust = 1))

ggsave(filename = "cutadapt_trimming_stats.pdf",
       device = "pdf",
       width = 6,
       height = 8,
       path = savedir
       )
```

## Contaminants

```{r bowtie2 container}
contaminant_files <- list.files(paste(wd,"bowtie2",sep="/"), 
  pattern = "contaminants.txt", 
  full.names = T, 
  recursive = T)

contaminant_df <- data.frame()

for (i in seq_along(contaminant_files)){
  contaminants <- read.delim(contaminant_files[i], skip = 1, header = T)[-1, ]
  colnames(contaminants)[1] <- "Sample"
  contaminants$Sample <- paste(
    strsplit(basename(contaminant_files[i]), "-.*")[[1]][1], 
    collapse = "-")

  contaminant_df <- rbind(contaminant_df, contaminants)
}

contaminant_df <- contaminant_df %>%
  dplyr::mutate(group = ifelse(Sample %in% c("150", "151", "152", "153", "154", "155", "156", "157", "158", "159", "160", "161", "162", "163", "164", "165", "166", "167", "168", "169"),"TUM","ORG"),
                READ_TYPE = factor(READ_TYPE, levels = c("tRNA",'mtDNA',"snoRNA","rRNA","snRNA","Passed"))
  )


```

```{r bowtie2 perc plot}
ggplot(
  data = contaminant_df, 
  aes(
    x = reorder(Sample, -READS), 
    fill = READ_TYPE, 
    y = READS)) +
  geom_bar(position = position_fill(), stat = "identity") +
  labs(
    fill = "Read type", 
    y = "Relative fraction of riboseq reads", 
    x = NULL) +
  ggsci::scale_fill_jco(alpha = 0.7) +
  theme_classic() +
  scale_y_continuous(expand = c(0, 0)) +
  theme(
    axis.ticks.x = element_blank(), 
    axis.line.x = element_blank(), 
    axis.text.x = element_text(angle = 45, hjust = 1), 
    axis.text = element_text(size = 6), 
    axis.text.y = element_text(size = 10),
    axis.title = element_text(size = 8)) +
  facet_grid(~ group, scales = "free_x")

ggsave(filename = "bowtie2_percentage_contaminant_stats.pdf",
       device = "pdf",
       path = savedir
       )

```

```{r bowtie2 plot}
ggplot(
  data = contaminant_df, 
  aes(
    x = reorder(Sample, -READS), 
    fill = READ_TYPE, 
    y = READS)) +
  geom_bar(position = position_stack(), stat = "identity") +
  labs(
    fill = "Read type", 
    y = "Number of riboseq reads", 
    x = NULL) +
  ggsci::scale_fill_jco(alpha = 0.7) +
  theme_classic() +
  scale_y_continuous(expand = c(0, 0)) +
  theme(
    axis.ticks.x = element_blank(), 
    axis.line.x = element_blank(), 
    axis.text.x = element_text(angle = 45, hjust = 1), 
    axis.text.y = element_text(size = 10,angle = 45),
    axis.text = element_text(size = 8))  +
  facet_grid(~ group, scales = "free_x")

ggsave(filename = "bowtie2_contaminant_stats.pdf",
       device = "pdf",
       path = savedir
       )
```

## Alignment

```{r STAR container}

star_stats <- read.delim(paste(wd,
                               "multiqc",
                               paste0(multiqc_name,"_multiqc_data"),
                               "multiqc_star.txt", sep ="/")) %>%
  dplyr::filter(!(grepl("STARpass1", Sample))) %>%
  dplyr::mutate(Sample = gsub("-L.*","",Sample),
                group = c(rep("ORG",31),rep("TUM",19)))

star_perc_stats <- star_stats %>%
  dplyr::select(c("Sample","group",
                  "uniquely_mapped_percent","multimapped_percent","multimapped_toomany_percent",
                  "unmapped_mismatches_percent",
                  "unmapped_tooshort_percent","unmapped_other_percent")) %>%
  dplyr::mutate(unmapped_other_percent = unmapped_other_percent + multimapped_toomany_percent + unmapped_mismatches_percent) %>%
  dplyr::select(-c("multimapped_toomany_percent",
                  "unmapped_mismatches_percent")) %>%
  tidyr::pivot_longer(cols = c(3:6)) %>%
  dplyr::mutate(Sample = factor(Sample, levels = star_stats[order(star_stats$uniquely_mapped_percent),]$Sample),
                name = factor(name,levels = c("unmapped_other_percent","unmapped_tooshort_percent",
                                              "multimapped_percent","uniquely_mapped_percent")))

star_stats <- star_stats %>%
  dplyr::select(c("Sample","group",
                  "uniquely_mapped", "multimapped", 
                  "unmapped_tooshort", "unmapped_other","multimapped_toomany","unmapped_mismatches")) %>%
  dplyr::mutate(unmapped_other = unmapped_other + multimapped_toomany + unmapped_mismatches) %>%
  dplyr::select(-c("multimapped_toomany",
                  "unmapped_mismatches")) %>%
  tidyr::pivot_longer(cols = c(3:6)) %>%
  dplyr::mutate(Sample = factor(Sample, levels = star_stats[order(star_stats$uniquely_mapped),]$Sample),
                name = factor(name,levels = c("unmapped_other","unmapped_tooshort",
                                              "multimapped","uniquely_mapped")))

```

```{r STAR plot}

ggplot(data = star_stats, aes(x = Sample, y = value, fill = name)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = rev(ggsci::pal_jco("default")(4))) +
  theme_classic() +
  labs(
       x = "Samples", 
       y = "Total RNAseq reads", 
       fill = "Read type") +
  theme(axis.text.x = element_text(angle = 65, hjust = 1)) +
  facet_grid(~ group, scales = "free_x")

ggsave(filename = "star_alignment_stats.pdf",
       device = "pdf",
       width = 8,
       height = 6,
       path = savedir
       )

```

```{r STAR perc plot}

ggplot(data = star_perc_stats, aes(x = Sample, y = value, fill = name)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = rev(ggsci::pal_jco("default")(4))) +
  theme_classic() +
  labs(x = "Samples", 
       y = "Total RNAseq reads", 
       fill = "Read type") +
  theme(axis.text.x  = element_text(angle = 65, hjust = 1))  +
  facet_grid(~ group, scales = "free_x")

ggsave(filename = "star_percentage_alignment_stats.pdf",
       device = "pdf",
       height = 6,
       width = 8,
       path = savedir
       )

```

## P sites

```{r riboseqc container}
riboseqc_files <- list.files(paste(wd,
                                   "RiboseQC",
                                   sep ="/"),
                             pattern = "results_RiboseQC_all", 
                             full.names = T, 
                             recursive = T)

riboseqc_files <- grep(pattern = "RMS_merged", riboseqc_files, invert = T, value = T)

# Create data.frame with all frame preferences
summary_P_sites_df <- data.frame()
summary_reads_df <- data.frame()
inframe_df <- data.frame()
read_cats_df <- data.frame()
cds_reads_df <- data.frame()
for (fname in riboseqc_files) {
  sample_id <- paste(strsplit(basename(fname), "-")[[1]][1], collapse = "-")
  message("Loading ", sample_id)
  load(fname)
  
  summary_P_sites_sample <- data.frame(res_all$summary_P_sites)
  summary_P_sites_sample$sample_id <- sample_id
  
  summary_reads_sample <- data.frame(t(colSums(data.frame(res_all$read_stats$reads_summary_unq$nucl))), row.names = sample_id)
  
  inframe_sample <- data.frame(t(res_all$selection_cutoffs$analysis_frame_cutoff$nucl$all$frames_res), row.names = sample_id)
  
  read_cats_sample <- data.frame(t(rowSums(data.frame(res_all$read_stats$reads_summary$nucl))), row.names = sample_id)

  cds_reads_sample <- data.frame(reads = t(sum(res_all$read_stats$counts_cds_genes_unq$reads)), row.names = sample_id)

  summary_P_sites_df <- rbind(summary_P_sites_df, summary_P_sites_sample)
  summary_reads_df <- dplyr::bind_rows(summary_reads_df, summary_reads_sample) 
  inframe_df <- rbind(inframe_df, inframe_sample)
  read_cats_df <- rbind(read_cats_df, read_cats_sample)
  cds_reads_df <- rbind(cds_reads_df, cds_reads_sample)
   # Clean up
  gc()
  rm(res_all)
}

summary_reads_df_sel <- summary_reads_df[,9:18]
perc_reads_df <- round((summary_reads_df_sel/rowSums(summary_reads_df_sel))*100,2)
colnames(perc_reads_df) <- gsub(colnames(perc_reads_df), pattern = ".*_", replacement = "")

read_cats_temp <- read_cats_df
read_cats_temp$sample_id <- rownames(read_cats_temp)
read_cats_long <- tidyr::pivot_longer(read_cats_temp, cols = colnames(read_cats_df))
read_cats_cds <- read_cats_long %>% 
  dplyr::filter(name == "cds") %>%
  dplyr::mutate(sample_id = factor(sample_id, levels = .[order(value),]$sample_id))

read_cats_long_df <- read_cats_long %>%
  dplyr::mutate(name = factor(name, levels = rev(c("cds","fiveutrs", "threeutrs", "ncRNAs", "ncIsof", "introns", "intergenic")))) %>%
  dplyr::group_by(sample_id) %>%
  dplyr::mutate(perc = round(value / sum(value)*100, 2),
                group = ifelse(sample_id %in% c("150", "151", "152", "153", "154", "155", "156", "157", "158", "159", "160", "161", "162", "163", "164", "165", "166", "167", "168", "169"),"TUM","ORG"),
                sample_id = factor(sample_id, levels = levels(read_cats_cds$sample_id)))

read_cats_annotgenes <- subset(read_cats_long_df, name %in% c("fiveutrs", "cds", "threeutrs"))

# Create plot DFs

cds_reads_df <- dplyr::mutate(cds_reads_df,group = c(rep("ORG",31),rep("TUM",19)))
frame_perc_plot <- perc_reads_df %>%
  dplyr::mutate(group = c(rep("ORG",31),rep("TUM",19))) %>%
  tidyr::pivot_longer(cols = colnames(.)[-11])
readlen_29_df <- subset(summary_P_sites_df, read_length == 29 & comp == "nucl") %>%
  dplyr::mutate(group = c(rep("ORG",31),rep("TUM",19)))

```


### RPF length

```{r riboseqc RPF length plot}

ggplot(frame_perc_plot,aes(y = value, x = name, col = group)) +
  ggbeeswarm::geom_beeswarm(size = 1, cex = 0.8, alpha = 1) +
  
  stat_summary(aes(group = 1, yend = ..y.., xend = ..x.. - 0.2), color = "black", fun = mean, geom = "segment") +
  stat_summary(aes(group = 1, yend = ..y.., xend = ..x.. + 0.2), color = "black", fun = mean, geom = "segment") +
  scale_y_continuous(labels = scales::percent_format(scale = 1)) +
  labs(y = "Reads", x = "Ribosome footprint length (nt)") +
  theme_classic() +facet_grid(~ group)

ggsave(filename = "ribo_rpf_length.pdf",
       path = savedir,
       width = 10,
       height = 5)
```

### inframe

```{r riboseqc inframe plot}
inframe_df$group <- c(rep("ORG",31),rep("TUM",19))

ggplot(data = inframe_df, aes(y = mean_pct_max_frame, 
                              x = reorder(rownames(inframe_df), -mean_pct_max_frame))) +
  geom_bar(stat = "identity", fill = colors[1]) +
  scale_y_continuous(labels = scales::percent_format(scale = 1), limits = c(0, 100), expand = c(0, 0)) +
  geom_hline(yintercept = mean(inframe_df$mean_pct_max_frame), linetype = 2) +
  labs(y = "In-frame reads (all read lengths)", x = "Samples") +
  theme_classic() +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 8)) +
    facet_grid(~ group, scales = "free_x")

ggsave(filename = "ribo_inframe_all_perc.pdf",
       path = savedir,
       width = 8,
       height = 5)
```

### inframe 29 nt

```{r riboseqc inframe 29nt plot}
ggplot(data = readlen_29_df, aes(y = frame_preference, x = reorder(sample_id, -frame_preference))) +
  geom_bar(stat = "identity", fill = colors[1]) +
  scale_y_continuous(labels = scales::percent_format(scale = 1), limits = c(0, 100), expand = c(0, 0)) +
  geom_hline(yintercept = mean(readlen_29_df$frame_preference), linetype = 2) +
  geom_hline(yintercept = 60, color = "red", linetype = 2) +

  labs(y = "In-frame reads (29nt)", x = "Samples") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 8)) + facet_grid(~ group, scales = "free_x")

ggsave(filename = "ribo_inframe_29nt_perc.pdf",
       path = savedir,
       width = 8,
       height = 5)
```

### Read categorisation

```{r riboseqc RPF location plot}
ggplot(data = read_cats_long_df,
       aes(y = perc, x = sample_id, fill = name)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = colors) +
  scale_y_continuous(labels = scales::percent_format(scale = 1),
                     limits = c(0, 101)) +
  labs(y = "Riboseq reads") +
  theme_classic() +
  theme(
    axis.ticks.x = element_blank(),
    axis.line.x = element_blank(),
    axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      size = 6)) +
  theme(
    strip.placement = "outside",
    strip.background = element_blank(),
    axis.title.x = element_blank()
  ) +
  facet_grid(~ group,
             scales = "free_x")
ggsave(filename = "ribo_read_loc.pdf",
       path = savedir,
       width = 10,
       height = 5)

```

### RPF categorisation

```{r riboseqc RPF location plot}
ggplot(data = read_cats_long,
       aes(y = perc, x = sample_id, fill = name)) +
  geom_bar(stat = "identity",
           position = "dodge2") +
  scale_fill_manual(values = colors) +
  scale_y_continuous(labels = scales::percent_format(scale = 1),
                     limits = c(0, 100)) +
  labs(y = "Riboseq reads") +
  theme_classic() +
  theme(
    axis.ticks.x = element_blank(),
    axis.line.x = element_blank(),
    legend.position = "none",
    axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      size = 3)) +
  facet_grid( 
    ~ name,
    scales = "free_x",  # Let the x axis vary across facets.
    space = "free_x",  # Let the width of facets vary and force all bars to have the same width.
    switch = "x") +
  
  theme(
    strip.placement = "outside",
    strip.background = element_blank(),
    axis.title.x = element_blank()
  )
ggsave(filename = "ribo_rpf_loc.pdf",
       path = savedir,
       width = 10,
       height = 5)

```

### CDS

```{r riboseqc CDS plot}

ggplot(data = cds_reads_df, aes(y = reads, x = reorder(rownames(cds_reads_df), -reads))) +
  geom_bar(stat = "identity", fill = colors[1]) +
  scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
  geom_hline(yintercept = mean(cds_reads_df$reads), linetype = 2) +
  labs(y = "CDS reads", x = "Samples") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 5)) + 
  facet_grid(~ group, scales = "free_x")

ggsave(filename = "ribo_CDS_count_barplot.pdf",
       path = savedir,
       width = 7,
       height = 4)
```

