---
title: "RMS combined ORFs"
author: "JD"
date: "2023-10-13"
output: html_document
---


TODO: annotated DF with protein sequence

```{r libraries}
library(dplyr)
library(stringr)
library(rtracklayer)
library(GenomicRanges)
library(magrittr)
library(data.table)
library(ggplot2)
```

```{r parameters}
savedir = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/results"
txome_gtf = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/01_rnaseq/analysis/rnaseq_pipeline/customannotation/RMS_full_novel_filtered_corrected.gtf"
txome_rannot = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/01_rnaseq/analysis/rnaseq_pipeline/customannotation/RMS_container/RMS_full_novel_filtered_corrected.gtf_Rannot"
orf_combined = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/results/combined/orfs_combined.RData"
orf_overlap = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/results/combined/orfs_overlap.RData"
orf_separate = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/results/combined/orfs_separate.RData"
sample_ids <- gsub("-.*","",grep("RMS",basename(list.dirs("/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/analysis/ORFquant", recursive = F)), value = T))
fasta_locs = grep("RMS",list.dirs("/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/analysis/ORFquant", recursive = F), value = T)
```

```{r functions}
filter_orfs <- function(m_overlap, m_orfs) {
  # Determine for each unique ORFs between how many samples it is shared
  # Parameters:
  #   m_overlap - matrix, created by get_partial_overlap function
  #   m_orfs - GRangesList, all ORFs from an ORF caller
  #
  # Returns:
  #   orfs_filtered - dataframe, rownames are ORF ids, the column indicates
  #     between how many samples the ORF was shared
  orfs_filtered <- m_overlap %>% 
    as.data.frame(row.names = names(m_orfs)) %>% 
    filter(`1_parent` == 0) %>% 
    dplyr::select(starts_with("1_") & !ends_with("parent")) %>% 
    apply(1, function(x) sum(x>0)) %>% 
    as.data.frame()
  names(orfs_filtered) <- "rowsum"
  return(orfs_filtered)
}
```

# Generate ORF data

This code block creates reference info for our little ORFs

```{r GTF based reference}
gtf_data <- data.table::fread(txome_gtf, skip = 5)
colnames(gtf_data) <- c("seqname", "source", "feature", "start", "end", 
                        "score", "strand", "frame", "attribute")
tid_gid <- gtf_data %>%
  dplyr::mutate(
    transcript_id = as.character(
      stringr::str_match(attribute, "ENST\\d+|MSTRG[.]\\d+[.]\\d+")
    ),
    gene_id = as.character(
      stringr::str_match(attribute, "ENST\\d+|MSTRG[.]\\d+[.]\\d+")
    )
  ) %>%
  dplyr::select(transcript_id, gene_id) %>%
  dplyr::filter(!is.na(transcript_id)) %>%
  dplyr::distinct()

gtf_exon <- gtf_data %>%
  dplyr::mutate(
    transcript_id = as.character(
      stringr::str_match(attribute, "ENST\\d+|MSTRG[.]\\d+[.]\\d+")
    ),
    gene_id = as.character(
      stringr::str_match(attribute, "ENST\\d+|MSTRG[.]\\d+[.]\\d+")
    )
  ) %>%
  dplyr::filter(feature == "exon") %>%
  dplyr::select(seqname, start, end, strand,
                transcript_id, gene_id)

lnc_genes <- gtf_data %>%
  dplyr::mutate(gene_id = as.character(
    stringr::str_match(attribute, "ENST\\d+|MSTRG[.]\\d+[.]\\d+")
  )) %>%
  dplyr::filter(feature == "gene") %>%
  dplyr::mutate(biotype = as.character(
    stringr::str_match(attribute, "(?<=gene_biotype \").*(?=\")")
  )) %>%
  dplyr::filter(biotype == "lncRNA") %>%
  dplyr::select(gene_id, biotype)

#Load GTF rannot file from ORFquant
load(txome_rannot)
gtf_ref <- GTF_annotation$cds_txs %>% as.data.frame() %>% 
    group_by(group_name) %>% 
    summarise(gtf_sta = min(start), gtf_end = max(end)) %>% 
    dplyr::rename(transcript_id = group_name)

```

First, we load the data into R

```{r load data}
# TXOME GTF
tx_gtf <- rtracklayer::import.gff(txome_gtf)
tx_df <- as.data.frame(tx_gtf)

# ORF data
load(orf_separate)
load(orf_combined)
load(orf_overlap)

### ORF matrix
# Merge ORFs with 80% overlap, only keep the single parent as a reference
orfs_keep <- filter_orfs(m_overlap = orfquant_overlap,
                      m_orfs = orfquant_orfs_combined) %>%
  dplyr::mutate(orf_id = row.names(.))

mat <- orfquant_overlap[,grepl("1_",colnames(orfquant_overlap))]
rownames(mat) <- names(orfquant_orfs_combined) 

mat <- as.data.frame(mat) %>%
  dplyr::mutate(orf_id = rownames(.)) %>%
  dplyr::filter(`1_parent` == "0") %>%
  dplyr::left_join(orfs_keep)
colnames(mat) <- c(sample_ids,"parent","orf_id","count")

```

Use the following DEseq2 result tables from the RNA-seq to filter for potential
RMS-specific genes

```{r Check RMS-specificity}

rms_canon <- read.delim("/hpc/pmc_vanheesch/projects/Jip/rms_analysis/01_rnaseq/results/quantification/RMS_canon_enriched_matrix.txt", sep = ";")

rms_canon_pass <- rms_canon %>% dplyr::filter(pass == T & selected == T)

rms_novel <- read.delim("/hpc/pmc_vanheesch/projects/Jip/rms_analysis/01_rnaseq/results/quantification/RMS_novel_enriched_matrix.txt", sep = ";")

rms_novel_pass <- rms_novel %>% dplyr::filter(pass == T)

rms_fp <- read.delim("/hpc/pmc_vanheesch/projects/Jip/rms_analysis/01_rnaseq/results/quantification/RMS_FP_res.txt", sep = ";")

rms_fp_pass <- rms_fp %>% dplyr::filter(log2FoldChange > 1 & padj < 0.01)

```

Stitch all the information together in a big table

```{r parse data into DF }

# Annotate the ORF IDs with gene-level data
orf_df <- data.frame(orf_id = names(orfquant_orfs_combined)) %>%
  dplyr::mutate(transcript_id = gsub("_.*","",names(orfquant_orfs_combined))) %>%
  dplyr::left_join(tx_df[tx_df$type == "transcript",
                         c("class_code","gene_name","gene_biotype","strand","transcript_id","gene_id")]) %>%
  dplyr::distinct()

# Create DF with ORF metadata
categories_all <- data.frame(mcols(orfquant_orfs_combined)) %>%
  dplyr::select(-gene_id,-transcript_id)

interesting_orfs <- orf_df %>%
  dplyr::left_join(categories_all, by = "orf_id") %>%
  # # Only use novel ORFs outside existing CDS
  # dplyr::filter(orf_type %in% c("dORF","novel","uORF") | 
  #                 gene_biotype %in% c("stringtie","lncRNA")) %>%
  dplyr::filter(
  # Find ORFs that occur in at least 2 samples
  orf_id %in% mat[which(mat$count > 1),"orf_id"]
  ) %>%
  
  # Append sample occurrence per ORF ID
  dplyr::left_join(mat[,c("orf_id","count")]) %>%
  
  # Annotate NBL-specificity
  dplyr::mutate(pass = ifelse(gene_id %in% c(rms_canon_pass$gene_id,
                                             rms_novel_pass$gene_id),T,F),
                fp = ifelse(gene_id %in% row.names(rms_fp_pass),T,F),
                source_name = c(98:127,129,150:161,163:169)[source]) %>%
  dplyr::distinct()

```

```{r write table}

write.table(interesting_orfs, file = paste(savedir,"combined","RMS_combined_all_ORFs.txt",sep="/"), quote = F, row.names = F, sep =";")


```

# ORF data

Create a new DF with a single row per ORF, aggregate P-value and P-site

```{r Parse ORF data}
interesting_orfs <- read.delim(paste(savedir,"combined","RMS_combined_all_ORFs.txt",sep="/"), sep =";")

single_orf <- interesting_orfs %>%
  dplyr::group_by(orf_id) %>%
  dplyr::summarise(median_psites = median(p_sites),
                   sum_psites = sum(p_sites),
                   mean_pval = mean(p_value)) %>%
  dplyr::left_join(interesting_orfs[,-c(9,10,12)]) %>%
  dplyr::mutate(orf_type = ifelse(gene_biotype == "lncRNA","lncORF",orf_type)) %>%
  dplyr::distinct()

```

# Generate FASTA

```{r tumoroid FASTA}
fasta_locs = grep("RMS",list.dirs("/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/analysis/ORFquant", 
                                  recursive = F), value = T)
savedir = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/results"
filtered_interesting_orfs = read.delim(paste(savedir,"combined","RMS_combined_all_ORFs.txt",sep="/"), sep =";")
savefile = paste(savedir,"combined","combined_sequence_table.txt", sep = "/")

# code to extract all proteins from the file(s)
processed_ids <- character(0)
matching_seq <- character(0)
sequence_table <- data.frame(orf_id = character(0), sequence = character(0), width = integer(0), names = character(0), stringsAsFactors = FALSE)

for (i in 1:nrow(filtered_interesting_orfs)) {
  id <- filtered_interesting_orfs$orf_id[i]
  source_file_number <- filtered_interesting_orfs$source[i]
  if (!(id %in% processed_ids)) {
    fasta_file <- list.files(grep(source_file_number,fasta_locs, value = T), pattern = "fasta", full.names = T)
    fasta <- Biostrings::readAAStringSet(fasta_file)
    matching_fasta <- fasta[grepl(id, names(fasta))]
    # Extract sequences, widths, and names
    matching_sequences <- as.character(matching_fasta)
    matching_widths <- width(matching_fasta)
    matching_names <- names(matching_fasta)
    # Add the extracted data to the sequence_table data frame
    if (length(matching_sequences) > 0) {
      new_rows <- data.frame(orf_id = id, sequence = matching_sequences, width = matching_widths, names = matching_names, stringsAsFactors = FALSE)
      sequence_table <- rbind(sequence_table, new_rows)
    }
    processed_ids <- c(processed_ids, id)
  }
}

patient_fa <- Biostrings::AAStringSet(sequence_table[,c("sequence")])
names(patient_fa) <- sequence_table[,c("names")]

Biostrings::writeXStringSet(patient_fa, filepath = paste(savedir,"combined","combined_ORFs.fa", sep = "/"),
                            format = "fasta")
write.table(x = sequence_table, file = savefile,quote=F,sep="\t", row.names = F)

```

## ORFquant loci

```{r ORFquant plot}
gene = "ENSG00000233369"
sample = "160"
orfquant_results = list.files(path = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/analysis/ORFquant/",
                             pattern = paste0(".*",sample,".*_final_ORFquant_results"), full.names = T, recursive = T)
bam_files = ""
plotfile=paste(savedir,paste0(gene,"_plot.pdf"), sep = "/")

load(orfquant_results)

plot_orfquant_locus(locus = gene,
                              orfquant_results = ORFquant_results,
                              bam_files = bam_files,
                              plotfile = plotfile)
```

Collect protein sequence from single gene

- Loaded in **interesting_orfs**
- 

```{r}

# Try to make it work for 1 ORF

orf = "MSTRG.31321.1_128_346"

samples_with_orf = interesting_orfs[which(interesting_orfs$orf_id == orf),"source"]

fasta = list.files(path = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/analysis/ORFquant/",
                   pattern = paste0(".*","165",".*_Protein_sequences.fasta"), full.names = T, recursive = T)

protein_seqs = Biostrings::readAAStringSet(fasta)
matching_indices <- grep(orf, names(protein_seqs), value = T)

# These all match
protein_seq_of_interest_a = protein_seqs[[matching_indices]]
protein_seq_of_interest_b = protein_seqs[[matching_indices]]
protein_seq_of_interest_c = protein_seqs[[matching_indices]]

```

```{r fasta seq}

# code to extract all proteins from the file(s)
processed_ids <- character(0)
matching_seq <- character(0)
sequence_table <- data.frame(orf_id = character(0), sequence = character(0), width = integer(0), names = character(0), stringsAsFactors = FALSE)
filtered_interesting_orfs <- interesting_orfs %>%
  replace(is.na(x = .), "c") %>%
  dplyr::filter(!(class_code == "k"))
for (i in 1:nrow(filtered_interesting_orfs)) {
  id <- filtered_interesting_orfs$orf_id[i]
  source_file_number <- filtered_interesting_orfs$source[i]
  if (!(id %in% processed_ids)) {
    fasta_file <- list.files(grep(source_file_number,fasta_locs, value = T), pattern = "fasta", full.names = T)
    fasta <- Biostrings::readAAStringSet(fasta_file)
    matching_fasta <- fasta[grepl(id, names(fasta))]
    # Extract sequences, widths, and names
    matching_sequences <- as.character(matching_fasta)
    matching_widths <- width(matching_fasta)
    matching_names <- names(matching_fasta)
    # Add the extracted data to the sequence_table data frame
    if (length(matching_sequences) > 0) {
      new_rows <- data.frame(orf_id = id, sequence = matching_sequences, width = matching_widths, names = matching_names, stringsAsFactors = FALSE)
      sequence_table <- rbind(sequence_table, new_rows)
    }
    processed_ids <- c(processed_ids, id)
  }
}

```

## Subset for MHC binding

* Only RMS-specific genes
* Flag FP+ specific genes
* Only run netMHCpan on these

Requires:
* interesting_orfs
* rms_fp_pass
* rms_novel_pass
* rms_canon_pass

```{r}

neoantigen_orfs <- interesting_orfs %>%
  dplyr::group_by(orf_id) %>%
  dplyr::summarise(median_psites = median(p_sites),
                   sum_psites = sum(p_sites),
                   mean_pval = mean(p_value)) %>%
  dplyr::left_join(interesting_orfs[,-c(9,10,12,16)]) %>%
  dplyr::mutate(orf_type = ifelse(gene_biotype == "lncRNA","lncORF",orf_type)) %>%
  dplyr::distinct() %>%
  dplyr::filter(gene_id %in% c(rms_canon_pass$gene_id,rms_novel_pass$gene_id)) %>%
  dplyr::left_join(interesting_orfs[,c("orf_id","p_sites","source","source_name")])

sel_orfs <- list()

split_orfs <- split(neoantigen_orfs, neoantigen_orfs$orf_id) %>%
  lapply(function(x) {
    max_psites = max(x$p_sites)
    sel_orf = x[x$p_sites == max_psites,]
    sel_orfs[x$orf_id] = sel_orf
  })

sel_orfs_df <- do.call(rbind,split_orfs)

savefile = paste(savedir,"combined","combined_neoantigen_strict_sequence_table.txt", sep = "/")

# code to extract all proteins from the file(s)
processed_ids <- character(0)
matching_seq <- character(0)
sequence_table <- data.frame(orf_id = character(0), sequence = character(0), width = integer(0), names = character(0), stringsAsFactors = FALSE)

for (i in 1:nrow(sel_orfs_df)) {
  id <- sel_orfs_df$orf_id[i]
  source_file_number <- sel_orfs_df$source[i]
  if (!(id %in% processed_ids)) {
    fasta_file <- list.files(fasta_locs[source_file_number], pattern = "fasta", full.names = T)
    fasta <- Biostrings::readAAStringSet(fasta_file)
    matching_fasta <- fasta[grepl(id, names(fasta))]
    # Extract sequences, widths, and names
    matching_sequences <- as.character(matching_fasta)
    matching_widths <- width(matching_fasta)
    matching_names <- names(matching_fasta)
    # Add the extracted data to the sequence_table data frame
    if (length(matching_sequences) > 0) {
      new_rows <- data.frame(orf_id = id, 
                             sequence = matching_sequences, 
                             width = matching_widths, 
                             names = matching_names, 
                             stringsAsFactors = FALSE)
      sequence_table <- rbind(sequence_table, new_rows)
    }
    processed_ids <- c(processed_ids, id)
  }
}

sequence_table$names <- gsub("\\|.*","",sequence_table[,c("names")])
                        
# Convert sequence table to fasta
fa <- Biostrings::AAStringSet(sequence_table[,c("sequence")])
names(fa) <- sequence_table$names

Biostrings::writeXStringSet(fa, filepath = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/netmhcpan_single/data/rms_neoantigen_strict.fasta",
                            format = "fasta")

```
