---
title: "RMS figures volcano"
author: "JD"
date: "2023-11-17"
output: html_document
---

# Volcano plot

```{r parameters}
basedir = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/01_rnaseq"
savedir = paste(basedir,"results/quantification/figures",sep="/")
metadata_dir = paste(basedir,"documentation",sep="/")
txome_gtf = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/01_rnaseq/analysis/rnaseq_pipeline/customannotation/RMS_full_novel_filtered_corrected.sorted.gtf"
tumor_type = "RMS"
```

```{r gene classes}
# Gene classes
tx_gtf <- rtracklayer::import.gff(txome_gtf)
tx_df <- as.data.frame(tx_gtf) %>%
  dplyr::select(gene_id,gene_biotype,gene_name) %>%
  dplyr::distinct() %>%
  dplyr::mutate(gene_name = ifelse(grepl("MSTRG",gene_id),gene_id,gene_name),
                gene_biotype = ifelse(grepl("pseudo",gene_biotype),"pseudogene_variant",gene_biotype))
```

## Patient

```{r load data}

rms_pat <- read.delim(file = paste(basedir,"results","quantification","RMS_patient_res.txt",sep = "/"), sep = ";") 

rms_pat <- rms_pat %>%
  dplyr::mutate(gene_id = rownames(.)) %>%
  dplyr::left_join(tx_df) %>%
  dplyr::filter(gene_biotype %in% c("stringtie","lncRNA","protein_coding","pseudogene_variant"))

```

```{r volcano general}
high_conf_genes <- rms_pat[which(-log10(rms_pat$padj) > 50),]$gene_id

rms_pat_plot <- rms_pat %>%
  dplyr::mutate(gene_label = gene_name,
                log2FoldChange = ifelse(log2FoldChange < -10,-10,
                                        ifelse(log2FoldChange > 10,10,log2FoldChange))) %>%
  dplyr::mutate(gene_label = ifelse(gene_id %in% high_conf_genes,gene_label,NA))

ggplot(rms_pat_plot, aes(y = -log10(padj), x = log2FoldChange, colour = gene_biotype, label = gene_label)) +
  geom_point(size = 1) +
  ggrepel::geom_text_repel(size = 3,show.legend = FALSE) +
  scale_x_continuous(limits = c(-10,10)) + 
  viridis::scale_color_viridis(discrete = T, end = 0.9) +
  theme_minimal() +
  geom_vline(xintercept=c(-0.6, 0.6), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red")
```

```{r volcano FP-target}
fp_genes <- fp_target_genes$gene_id

highlighted_genes <- c()

rms_pat_enr <- rms_pat %>%
  dplyr::filter(gene_id %in% fp_genes) %>%
  dplyr::mutate(gene_label = gene_name) %>%
  dplyr::left_join(fp_target_genes[,c("gene_id","regulation","PF_target")]) %>%
  dplyr::mutate(gene_label = ifelse(is.na(PF_target),gene_label,NA))

ggplot(rms_pat_enr, aes(y = -log10(padj), x = log2FoldChange, colour = regulation,label=gene_label)) +
  geom_point(size = 1) +
  ggrepel::geom_text_repel(size = 3,show.legend = FALSE) +
  scale_x_continuous() + 
  viridis::scale_color_viridis(discrete = T, end = 0.9) +
  theme_minimal() +
  geom_vline(xintercept=c(-0.6, 0.6), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red")


```

```{r volcano enriched FP-target}

rms_pat_enr <- rms_pat %>%
  dplyr::filter(gene_id %in% fp_genes & gene_id %in% c(fp_pass_canon[which(fp_pass_canon$pass == T),]$gene_id,
                                                       fn_pass_canon[which(fn_pass_canon$pass == T),]$gene_id,
                                                       combined_pass_canon[which(combined_pass_canon$pass == T),]$gene_id)) %>%
  dplyr::mutate(gene_label = gene_name) %>%
  dplyr::left_join(fp_target_genes[,c("gene_id","regulation","PF_target")])

ggplot(rms_pat_enr, aes(y = -log10(padj), x = log2FoldChange, colour = regulation,label=gene_label)) +
  geom_point(size = 1) +
  ggrepel::geom_text_repel(size = 3,show.legend = FALSE) +
  scale_x_continuous() + 
  viridis::scale_color_viridis(discrete = T, end = 0.7) +
  theme_minimal() +
  geom_vline(xintercept=c(-0.6, 0.6), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red")

```

## Tumoroid

```{r load data}
rms_tum <- read.delim(file = paste(basedir,"results","quantification","tumoroid","RMS_FP_tumoroid_res.txt",sep = "/"), sep = ";") 

rms_tum <- rms_tum %>%
  dplyr::mutate(gene_id = rownames(.)) %>%
  dplyr::left_join(tx_df) %>%
  dplyr::filter(gene_biotype %in% c("stringtie","lncRNA","protein_coding","pseudogene_variant"))
  
```

```{r volcano FP-target}
fp_genes <- fp_target_genes$gene_id

highlighted_genes <- c()

rms_tum_enr <- rms_tum %>%
  dplyr::filter(gene_id %in% fp_genes) %>%
  dplyr::mutate(gene_label = gene_name) %>%
  dplyr::left_join(fp_target_genes[,c("gene_id","regulation","PF_target")]) %>%
  dplyr::mutate(gene_label = ifelse(is.na(PF_target),gene_label,NA))

ggplot(rms_tum_enr, aes(y = -log10(padj), x = log2FoldChange, colour = regulation,label=gene_label)) +
  geom_point(size = 1) +
  ggrepel::geom_text_repel(size = 3,show.legend = FALSE) +
  scale_x_continuous() + 
  viridis::scale_color_viridis(discrete = T, end = 0.9) +
  theme_minimal() +
  geom_vline(xintercept=c(-0.6, 0.6), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red")
```

```{r volcano enriched FP-target}
fp_genes <- fp_target_genes$gene_id

highlighted_genes <- c()

rms_tum_enr <- rms_tum %>%
  dplyr::filter(gene_id %in% fp_genes & gene_id %in% c(fp_pass_canon[which(fp_pass_canon$pass == T),]$gene_id,
                                                       fn_pass_canon[which(fn_pass_canon$pass == T),]$gene_id,
                                                       combined_pass_canon[which(combined_pass_canon$pass == T),]$gene_id)) %>%
  dplyr::mutate(gene_label = gene_name) %>%
  dplyr::left_join(fp_target_genes[,c("gene_id","regulation","PF_target")])

ggplot(rms_tum_enr, aes(y = -log10(padj), x = log2FoldChange, colour = regulation,label=gene_label)) +
  geom_point(size = 1) +
  ggrepel::geom_text_repel(size = 3,show.legend = FALSE) +
  scale_x_continuous() + 
  viridis::scale_color_viridis(discrete = T, end = 0.7) +
  theme_minimal() +
  geom_vline(xintercept=c(-0.6, 0.6), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red")
```
