---
title: "Merging RiboseQC P-site outputs"
output: html_notebook
---

In this notebook, I merged the `for_ORFquant` RiboseQC outputs of all patient samples + cell lines / organoid samples into a single merged `for_ORFquant` file to be used for a merged ORFquant run. 
This might be preferable to running ORFquant on a merged BAM file, since running RiboseQC individually on each sample will result in a list of P-sites based on the cutoffs selected for that sample, whereas a merged BAM file will be a combination of different samples with different optimal cutoffs. 

```{r parameters}

workdir <- "/hpc/pmc_vanheesch/projects/jvandinter/rms_analysis/02_riboseq/analysis"
savefile <- paste(workdir,"merged_p_sites/for_ORFquant_merged", sep = "/")
```

```{r}
library(IRanges)
library(ORFquant)
library(dplyr)
library(GenomicRanges)

# Function to process and merge P sites from GRanges objects
processAndMergePSites <- function(p_sites_list) {
  p_sites_plus <- p_sites_list[strand(p_sites_list) == "+"]
  p_sites_min <- p_sites_list[strand(p_sites_list) == "-"]

  cov_plus <- coverage(p_sites_plus, weight = mcols(p_sites_plus)$score)
  cov_plus <- GRanges(cov_plus)
  cov_plus <- cov_plus[cov_plus$score > 0]

  cov_min <- coverage(p_sites_min, weight = mcols(p_sites_min)$score)
  cov_min <- GRanges(cov_min)
  cov_min <- cov_min[cov_min$score > 0]

  strand(cov_plus) <- "+"
  strand(cov_min) <- "-"
  
  p_sites_merged <- sort(c(cov_plus, cov_min))
  rm(p_sites_plus, p_sites_min, cov_plus, cov_min)
  return(p_sites_merged)
}

# Function to merge junctions (summing reads and unique_reads)
mergeJunctions <- function(junctions_list) {
  summed_reads <- mcols(junctions_list[[1]])$reads
  summed_unique_reads <- mcols(junctions_list[[1]])$unique_reads
  
  for (i in 2:length(junctions_list)) {
    summed_reads <- summed_reads + mcols(junctions_list[[i]])$reads
    summed_unique_reads <- summed_unique_reads + mcols(junctions_list[[i]])$unique_reads
  }
  
  junctions_merged <- junctions_list[[1]]
  mcols(junctions_merged)$reads <- summed_reads
  mcols(junctions_merged)$unique_reads <- summed_unique_reads
  
  return(junctions_merged)
}


# Load for_ORFquant files
for_orfquant_files <- list.files(paste(workdir,"RiboseQC",sep = "/"), 
                                 pattern = "for_ORFquant", 
                                 recursive = TRUE, 
                                 full.names = TRUE)

for_orfquant_files <- grep(pattern = "RMS_merged", for_orfquant_files, invert = T, value = T)

p_sites_all_list <- GRangesList()
p_sites_uniq_list <- GRangesList()
p_sites_uniq_mm_list <- GRangesList()
junctions_list <- list()

for (i in seq_along(for_orfquant_files)) {
  print(basename(for_orfquant_files[i]))
  for_orfquant_file <- get(load(for_orfquant_files[i])) # Assuming this loads for_orfquant_file
  p_sites_all_list[[i]] <- for_orfquant_file$P_sites_all
  p_sites_uniq_list[[i]] <- for_orfquant_file$P_sites_uniq
  p_sites_uniq_mm_list[[i]] <- for_orfquant_file$P_sites_uniq_mm
  junctions_list[[i]] <- for_orfquant_file$junctions
  # Clean up
  rm(for_orfquant_file)
  rm(for_ORFquant)
  gc()
}

# Unlist, process, and merge for each P sites type
merged_p_sites_all <- processAndMergePSites(unlist(p_sites_all_list, recursive = FALSE))
merged_p_sites_uniq <- processAndMergePSites(unlist(p_sites_uniq_list, recursive = FALSE))
merged_p_sites_uniq_mm <- processAndMergePSites(unlist(p_sites_uniq_mm_list, recursive = FALSE))
merged_junctions <- mergeJunctions(junctions_list)

for_ORFquant <- list()
for_ORFquant$P_sites_all <- merged_p_sites_all
for_ORFquant$P_sites_uniq <- merged_p_sites_uniq
for_ORFquant$P_sites_uniq_mm <- merged_p_sites_uniq_mm
for_ORFquant$junctions <- merged_junctions

save(for_ORFquant, file = savefile)
```

