---
title: "RMS patient ORFs"
author: "JD"
date: "2023-08-28"
output: html_document
---

TODO: annotated DF with protein sequence

```{r libraries}
library(dplyr)
library(stringr)
library(rtracklayer)
library(GenomicRanges)
library(magrittr)
library(data.table)
library(ggplot2)
```

```{r parameters}
savedir = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/results"
txome_gtf = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/01_rnaseq/analysis/rnaseq_pipeline/customannotation/RMS_full_novel_filtered_corrected.gtf"
txome_rannot = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/01_rnaseq/analysis/rnaseq_pipeline/customannotation/RMS_container/RMS_full_novel_filtered_corrected.gtf_Rannot"
orf_combined = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/results/tumoroid/tumoroid_orfs_combined.RData"
orf_overlap = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/results/tumoroid/tumoroid_orfs_overlap.RData"
orf_separate = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/results/tumoroid/tumoroid_orfs_separate.RData"
sample_ids <- gsub("-.*","",grep("ORG",basename(list.dirs("/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/analysis/ORFquant", recursive = F)), value = T))
fasta_locs = grep("ORG",list.dirs("/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/analysis/ORFquant", recursive = F), value = T)
```

```{r functions}

plot_orfquant_locus <-
  function(locus,
           orfquant_results,
           bam_files,
           plotfile = 'locusplot.pdf',
           col = 'green') {
    if (!requireNamespace(c("Gviz", 'lemon', 'dplyr'), quietly = TRUE)) {
      stop(
        "Packages \"Gviz\",\"dplyr\",\"lemon\" needed for this function to work. Please install it.",
        call. = FALSE
      )
    }
    options(ucscChromosomeNames = FALSE)
    `%>%` <- dplyr::`%>%`
    
    if (is.null(orfquant_results$psite_data_file)) {
      stop(
        "this object looks like it's form an old version of ORFquant, it doesn't list the psite data file"
      )
    }
    if (length(orfquant_results$psite_data_file) > 1) {
      stop(
        "locus plots aren't supported for multiple psite tracks - either unify the psite tracks or modify the input object to have only one track"
      )
    }
    riboseqcoutput <- get(load(orfquant_results$psite_data_file))
    
    anno <- GTF_annotation
    
    
    selgene <- locus
    seltxs <-
      orfquant_results$ORFs_tx[orfquant_results$ORFs_tx$gene_id == selgene]$transcript_id
    selorfs <-
      orfquant_results$ORFs_tx[orfquant_results$ORFs_tx$gene_id == selgene]$ORF_id_tr
    stopifnot(length(locus) == 1)
    stopifnot(any(orfquant_results$ORFs_tx$gene_id == selgene))
    orfs_quantified_gen <-  orfquant_results$ORFs_gen[selorfs]
    
    mcols(orfs_quantified_gen) <-
      mcols(orfquant_results$ORFs_tx)[match(names(orfs_quantified_gen),
                                            orfquant_results$ORFs_tx$ORF_id_tr), ]
    seltxs <- orfs_quantified_gen$transcript_id %>% unique
    orfs_quantified_gen$feature <- 'CDS'
    orfs_quantified_gen$transcript = orfs_quantified_gen$transcript_id
    anno <- GTF_annotation
    orfs_quantified_tr <- anno$exons_tx
    orfs_quantified_tr <-
      orfs_quantified_tr[unique(orfs_quantified_gen$transcript_id)]
    #get the orfs, then get the negative coverage
    #add transcript info to the ORFs_tx object
    seqinf <-
      Seqinfo(names(anno$exons_tx), anno$exons_tx %>% width %>% sum)
    #Now get the negatives for each ORF
    
    utrs <-
      orfquant_results$ORFs_tx %>% subset(gene_id == selgene) %>% keepSeqlevels(seltxs) %>%
      {
        seqinfo(.) <- seqinf[seltxs]
        .
      } %>%
      coverage %>%
      as('GRanges') %>% subset(score == 0) %>% mapFromTranscripts(anno$exons_tx) %>%
      {
        .$transcript <- names(anno$exons_tx)[.$transcriptsHits]
        .
      } %>%
      {
        .$feature = 'utr'
        .
      }
    
    orfquantgr <- c(orfs_quantified_gen[, c('feature', 'transcript')]
                    , utrs[, c('feature', 'transcript')])
    orfquantgr$feature[orfquantgr$feature == 'CDS'] <-
      names(orfquantgr)[orfquantgr$feature == 'CDS']
    #get correct col name
    metacols = colnames(mcols(orfs_quantified_gen))
    quantcol = 'ORFs_pM'
    
    ufeats <- orfquantgr$feature %>% {
      . = .[. != 'utr']
      .
    } %>% unique
    orfscores <-
      ufeats %>% setNames(match(ufeats, orfs_quantified_gen$ORF_id_tr) %>% {
        mcols(orfs_quantified_gen[.])[[quantcol]]
      }, .)
    
    
    orfcols <- orfscores %>% {
      . / max(na.omit(.))
    } %>%
      c(0, .)
    
    orfquantgrscores = mcols(orfs_quantified_gen)[[quantcol]][match(names(orfquantgr), orfs_quantified_gen$ORF_id_tr)]
    
    orfcols <-
      orfquantgrscores %>% vapply(function(.)
        tryCatch({
          rgb(0, ., 0)
        }, error = function(e) {
          'white'
        }), 'foo') %>% setNames(c('0', orfquantgr$feature %>% unique))
    
    
    ###Define non selected
    disctxs <-
      anno$txs_gene[selgene] %>% unlist %>% .$tx_name %>% unique %>% setdiff(seltxs)
    disctxsint <-
      disctxs %>% intersect(seqnames(anno$cds_txs_coords)) %>% as.character
    disc_orfquantgr <- anno$cds_txs_coords %>%
      keepSeqlevels(disctxsint, 'coarse') %>%
      {
        seqinfo(.) <- seqinf[disctxsint]
        .
      } %>%  coverage %>%
      as('GRanges') %>%
      subset(.$score == 0) %>%
      {
        txgr = .
        out = mapFromTranscripts(txgr, anno$exons_tx)
        out$score = txgr$score[out$xHits]
        out
      } %>%
      {
        .$transcript <- names(anno$exons_tx)[.$transcriptsHits]
        .
      } %>%
      {
        .$feature = ifelse(.$score == 0, 'utr', 'CDS')
        .
      }
    disc_orfquantgr <-
      disc_orfquantgr %>% c(., anno$cds_txs[disctxsint] %>% unlist %>% {
        .$feature = rep('CDS', length(.))
        .$transcript = names(.)
        .
      })
    discORFnames <-
      paste0(disctxs,
             '_',
             start(anno$cds_txs_coords[disctxsint]),
             '_',
             end(anno$cds_txs_coords[disctxsint])) %>% setNames(disctxsint)
    disc_orfquantgr$symbol = discORFnames[disc_orfquantgr$transcript]
    fakejreads <-
      riboseqcoutput$junctions %>% subset(any(gene_id == selgene)) %>% resize(width(.) +
                                                                                2, 'center') %>%
      {
        .$cigar <- paste0('1M', width(.) - 2, 'N', '1M')
        .
      }
    fakejreads <-
      fakejreads[mapply(seq_along(fakejreads[]), fakejreads$reads, FUN = rep) %>%
                   unlist]
    ncols <- 2
    nrows <- 1
    orfcols <- orfcols[order(-orfscores[names(orfcols)])]
    orfquantgr_sorted <-
      orfquantgr[order(orfscores[names(orfquantgr)])]
    fix_utrs <- function(orfquantgr_sorted) {
      orfquantgr_sorted$symbol = names(orfquantgr_sorted)
      #add utrs for each ORF
      #for each selected ORF
      orftrpairs <-
        orfquantgr_sorted %>% subset(feature != 'utr') %>% mcols %>% as.data.frame %>%
        {
          dplyr::distinct(.)
        }
      orfutrs <- orfquantgr_sorted %>% subset(feature == 'utr')
      orfutrs <- lapply(1:nrow(orftrpairs), function(i) {
        orfutrs <- orfutrs %>% subset(transcript == orftrpairs$transcript[i])
        orfutrs$symbol = orftrpairs$feature[i]
        names(orfutrs) = orfutrs$symbol
        orfutrs
      }) %>% GRangesList %>% unlist
      orfquantgr_sorted <-
        orfquantgr_sorted %>% subset(feature != 'utr') %>% c(., orfutrs)
      orfquantgr_sorted
    }
    orfquantgr_sorted <- fix_utrs(orfquantgr_sorted)
    # disc_orfquantgrfix<-fix_utrs(disc_orfquantgr)
    disc_orfquantgrfix <- (disc_orfquantgr)
    #dimenions, extent of the plot
    selgenerange <-  anno$genes[selgene]
    plotstart = start(selgenerange) - (0.2 * (end(selgenerange) - start(selgenerange)))
    plotend = end(selgenerange) + (0 * (end(selgenerange) - start(selgenerange)))

    legendwidth = 1 / 10
    plottitle <- paste0('ORFquant: ', selgene)
    #write to pdf
    pdf(plotfile, width = 14 + 2, h = 7)
    #code for arranging legend next to the locus plot
    grid.newpage()
    vp1 <- viewport(
      x = 0,
      y = 0,
      w = 1 - legendwidth * 1.5,
      h = 1,
      just = c("left", "bottom"),
      name = "vp1"
    )
    vp2 <-
      viewport(
        x = 1 - legendwidth * 1.5,
        y = 0,
        w = legendwidth * 1.5,
        h = 1,
        just = c("left", "bottom")
      )
    vp3 <-
      viewport(
        x = 1 - legendwidth * 1.5,
        y = 0,
        w = legendwidth * 1.5,
        h = 1 / 7,
        just = c("left", "bottom")
      )
    pushViewport(vp1)
    #finally plot the locus
    Gviz::plotTracks(
      main = plottitle,
      cex.main = 2,
      legend = TRUE,
      add = TRUE,
      from = plotstart,
      to = plotend,
      #zoomed in on the orf in question
      sizes = c(1, 1, 1, 1, 1, 1),
      rot.title = 0,
      cex.title = 1,
      title.width = 2.5,
      c(
        Gviz::GenomeAxisTrack(range = selgenerange),
        # Gviz::rnaseqtrack, # plot the riboseq signal
        # Gviz::txs_discarded_Track,
        # Gviz::txs_selected_track,
        # Gviz::DataTrack(riboseqcoutput$P_sites_all%>%subsetByOverlaps(selgenerange),type='hist'),
        Gviz::GeneRegionTrack(
          name = 'discarded\ntranscripts',
          anno$exons_tx[disctxs] %>% unlist %>% {
            .$transcript = names(.)
            .$feature = rep('exon', length(.))
            .
          },
          fill = '#F7CAC9',
          transcriptAnnotation = 'transcript'
        ),
        Gviz::GeneRegionTrack(
          exon = 'forestgreen',
          name = 'selected\ntranscripts',
          anno$exons_tx[seltxs] %>% unlist %>% {
            .$transcript = names(.)
            .$feature = rep('exon', length(.))
            .
          },
          fill = '#F7CAC9',
          transcriptAnnotation = 'transcript'
        ),
        Gviz::DataTrack(
          legend = TRUE,
          name = '\t\t P-Sites',
          col.histogram = 'forestgreen',
          riboseqcoutput$P_sites_all %>% subsetByOverlaps(selgenerange),
          type = 'hist'
        ),
        #Gviz::GeneRegionTrack(discarded_orfs_gen),
        Gviz::GeneRegionTrack(
          name = 'Discarded\nORFs',
          disc_orfquantgrfix,
          transcriptAnnotation = 'symbol',
          collapse = FALSE,
          thinBoxFeature = 'utr',
          CDS = 'blue',
          utr = 'white'
        ),
        Gviz::GeneRegionTrack(
          name = 'Selected\nORFs',
          range = orfquantgr_sorted,
          collapse = FALSE,
          thinBoxFeature = 'utr',
          CDS = 'red',
          utr = 'white',
          transcriptAnnotation = 'symbol'
        ) %>%
          # identity
          {
            displayPars(.)[names(orfcols)] <- orfcols
            .
          }
      ),
      col.labels = 'black',
      chr = seqnames(selgenerange)
    )
    #create barchart of intensities
    popViewport(1)
    pushViewport(vp2)
    cols = I(c(orfcols[which.min(orfscores)], orfcols[which.max(orfscores)]))
    grid.draw(lemon::g_legend(
      qplot(
        x = 1:2,
        y = 1:2,
        color = range(orfscores, na.rm = T)
      ) +
        scale_color_gradient(
          name = 'Normalized ORF Expr\n(ORFs_pM)',
          breaks = setNames(
            sort(na.omit(orfscores)),
            floor(na.omit(sort(orfscores))) %>% format(big.mark = ",", scientific =
                                                         FALSE)
          ),
          low = cols[1],
          high = cols[2]
        ) + theme(
          text = element_text(size = 14),
          legend.key.size = unit(.5, 'inches')
        )
    ))
    popViewport(1)
    pushViewport(vp3)
    dev.off()
    normalizePath(plotfile)
    #return file name
    return(plotfile)
  } 


filter_orfs <- function(m_overlap, m_orfs) {
  # Determine for each unique ORFs between how many samples it is shared
  # Parameters:
  #   m_overlap - matrix, created by get_partial_overlap function
  #   m_orfs - GRangesList, all ORFs from an ORF caller
  #
  # Returns:
  #   orfs_filtered - dataframe, rownames are ORF ids, the column indicates
  #     between how many samples the ORF was shared
  orfs_filtered <- m_overlap %>% 
    as.data.frame(row.names = names(m_orfs)) %>% 
    filter(`1_parent` == 0) %>% 
    dplyr::select(starts_with("1_") & !ends_with("parent")) %>% 
    apply(1, function(x) sum(x>0)) %>% 
    as.data.frame()
  names(orfs_filtered) <- "rowsum"
  return(orfs_filtered)
}
```

This code block creates reference info for our little ORFs

```{r GTF based reference}
gtf_data <- data.table::fread(txome_gtf, skip = 5)
colnames(gtf_data) <- c("seqname", "source", "feature", "start", "end", 
                        "score", "strand", "frame", "attribute")
tid_gid <- gtf_data %>%
  dplyr::mutate(
    transcript_id = as.character(
      stringr::str_match(attribute, "ENST\\d+|MSTRG[.]\\d+[.]\\d+")
    ),
    gene_id = as.character(
      stringr::str_match(attribute, "ENST\\d+|MSTRG[.]\\d+[.]\\d+")
    )
  ) %>%
  dplyr::select(transcript_id, gene_id) %>%
  dplyr::filter(!is.na(transcript_id)) %>%
  dplyr::distinct()

gtf_exon <- gtf_data %>%
  dplyr::mutate(
    transcript_id = as.character(
      stringr::str_match(attribute, "ENST\\d+|MSTRG[.]\\d+[.]\\d+")
    ),
    gene_id = as.character(
      stringr::str_match(attribute, "ENST\\d+|MSTRG[.]\\d+[.]\\d+")
    )
  ) %>%
  dplyr::filter(feature == "exon") %>%
  dplyr::select(seqname, start, end, strand,
                transcript_id, gene_id)

lnc_genes <- gtf_data %>%
  dplyr::mutate(gene_id = as.character(
    stringr::str_match(attribute, "ENST\\d+|MSTRG[.]\\d+[.]\\d+")
  )) %>%
  dplyr::filter(feature == "gene") %>%
  dplyr::mutate(biotype = as.character(
    stringr::str_match(attribute, "(?<=gene_biotype \").*(?=\")")
  )) %>%
  dplyr::filter(biotype == "lncRNA") %>%
  dplyr::select(gene_id, biotype)

#Load GTF rannot file from ORFquant
load(txome_rannot)
gtf_ref <- GTF_annotation$cds_txs %>% as.data.frame() %>% 
    group_by(group_name) %>% 
    summarise(gtf_sta = min(start), gtf_end = max(end)) %>% 
    dplyr::rename(transcript_id = group_name)

```

First, we load the data into R

```{r load data}
# TXOME GTF
tx_gtf <- rtracklayer::import.gff(txome_gtf)
tx_df <- as.data.frame(tx_gtf)

# ORF data
load(orf_separate)
load(orf_combined)
load(orf_overlap)

### ORF matrix
# Merge ORFs with 80% overlap, only keep the single parent as a reference
orfs_keep <- filter_orfs(m_overlap = orfquant_overlap,
                      m_orfs = orfquant_orfs_combined) %>%
  dplyr::mutate(orf_id = row.names(.))

mat <- orfquant_overlap[,grepl("^1_",colnames(orfquant_overlap))]
rownames(mat) <- names(orfquant_orfs_combined) 

mat <- as.data.frame(mat) %>%
  dplyr::mutate(orf_id = rownames(.)) %>%
  dplyr::filter(`1_parent` == "0") %>%
  dplyr::left_join(orfs_keep)
colnames(mat) <- c(sample_ids,"parent","orf_id","count")

```

Use the following DEseq2 result tables from the RNA-seq to filter for potential
RMS-specific genes

```{r Check RMS-specificity}

rms_canon <- read.delim("/hpc/pmc_vanheesch/projects/Jip/rms_analysis/01_rnaseq/results/quantification/RMS_canon_enriched_matrix.txt", sep = ";")

rms_canon_pass <- rms_canon %>% dplyr::filter(pass == T & selected == T)

rms_novel <- read.delim("/hpc/pmc_vanheesch/projects/Jip/rms_analysis/01_rnaseq/results/quantification/RMS_novel_enriched_matrix.txt", sep = ";")

rms_novel_pass <- rms_novel %>% dplyr::filter(pass == T)

rms_fp <- read.delim("/hpc/pmc_vanheesch/projects/Jip/rms_analysis/01_rnaseq/results/quantification/RMS_FP_res.txt", sep = ";")

rms_fp_pass <- rms_fp %>% dplyr::filter(log2FoldChange > 1 & padj < 0.01)

```

Stitch all the information together in a big table

```{r parse data into DF }

# Annotate the ORF IDs with gene-level data
orf_df <- data.frame(orf_id = names(orfquant_orfs_combined)) %>%
  dplyr::mutate(transcript_id = gsub("_.*","",names(orfquant_orfs_combined))) %>%
  dplyr::left_join(tx_df[tx_df$type == "transcript",
                         c("class_code","gene_name","gene_biotype","strand","transcript_id","gene_id")]) %>%
  dplyr::distinct()

# Create DF with ORF metadata
categories_all <- data.frame(mcols(orfquant_orfs_combined)) %>%
  dplyr::select(-gene_id,-transcript_id)

interesting_orfs <- orf_df %>%
  dplyr::left_join(categories_all, by = "orf_id") %>%
  # # Only use novel ORFs outside existing CDS
  # dplyr::filter(orf_type %in% c("dORF","novel","uORF") | 
  #                 gene_biotype %in% c("stringtie","lncRNA")) %>%
  dplyr::filter(
  # Find ORFs that occur in at least 2 samples
  orf_id %in% mat[which(mat$count > 1),"orf_id"]
  ) %>%
  
  # Append sample occurrence per ORF ID
  dplyr::left_join(mat[,c("orf_id","count")]) %>%
  
  # Annotate NBL-specificity
  dplyr::mutate(pass = ifelse(gene_id %in% c(rms_canon_pass$gene_id,
                                             rms_novel_pass$gene_id),T,F),
                fp = ifelse(gene_id %in% row.names(rms_fp_pass),T,F),
                source = sample_ids[source]) %>%
  dplyr::distinct()

```

```{r}

test_orfs <- orf_df %>%
  dplyr::left_join(categories_all, by = "orf_id") %>%
  dplyr::filter(
  # Find ORFs that occur in at least 2 samples
  orf_id %in% mat[which(mat$count > 1),"orf_id"]
  ) %>%
  
  # Append sample occurrence per ORF ID
  dplyr::left_join(mat[,c("orf_id","count")]) %>%
  
  # Annotate NBL-specificity
  dplyr::mutate(pass = ifelse(gene_id %in% c(rms_canon_pass$gene_id,
                                             rms_novel_pass$gene_id),T,F),
                fp = ifelse(gene_id %in% row.names(rms_fp_pass),T,F),
                source = c(150:161,163:169)[source]) %>%
  dplyr::distinct()

```

```{r write table}

write.table(interesting_orfs, file = paste(savedir,"tumoroid","RMS_tumoroid_all_ORFs.txt",sep="/"), quote = F, row.names = F, sep =";")


```

Create a new DF with a single row per ORF, aggregate P-value and P-site

```{r Parse ORF data}
interesting_orfs <- read.delim(paste(savedir,"tumoroid","RMS_tumoroid_all_ORFs.txt",sep="/"), sep =";")

single_orf <- interesting_orfs %>%
  dplyr::group_by(orf_id) %>%
  dplyr::summarise(median_psites = median(p_sites),
                   sum_psites = sum(p_sites),
                   mean_pval = mean(p_value)) %>%
  dplyr::left_join(interesting_orfs[,-c(9,10,12)]) %>%
  dplyr::mutate(orf_type = ifelse(gene_biotype == "lncRNA","lncORF",orf_type)) %>%
  dplyr::distinct()

```

# PLOTS

## ORF count distribution

```{r plot }
single_orf %>% dplyr::mutate(orf_type = ifelse(!(orf_type %in% c("uORF","novel","lncORF","dORF","ORF_annotated")),
                                               "other",orf_type)
  dplyr::mutate(orf_type = factor(orf_type, levels = c("uORF","ORF_annotated","dORF","novel","lncORF","other"))) %>%
  ggplot(aes(x = count, fill = orf_type)) +
  geom_histogram(stat = "count") +
  scale_x_continuous(breaks = length(sample_ids)) +
  facet_wrap(~ orf_type, scales = "free_y") +
  theme_classic()

single_orf %>% dplyr::mutate(orf_type = factor(orf_type, levels = c("uORF","novel","lncORF","dORF","ORF_annotated"))) %>%
  ggplot(aes(x = count, fill = orf_type)) +
  geom_density() +
  scale_x_continuous(breaks = length(sample_ids)) +
  facet_wrap(~ orf_type, scales = "free_y") +
  theme_classic()

ggsave(filename = "",
       device = "pdf",
       height =,
       width =,
       path = savedir)

single_orf %>% dplyr::filter(pass == T)  %>%
  dplyr::mutate(orf_type = factor(orf_type, levels = c("uORF","novel","lncORF","dORF","ORF_annotated"))) %>%
  ggplot(aes(x = count, fill = orf_type)) +
  geom_histogram(stat = "count") +
  scale_x_continuous(breaks = 2:length(sample_ids)) +
  facet_wrap(~ orf_type, scales = "free_y") +
  theme_classic()

single_orf %>% dplyr::filter(fp == T & pass == T)  %>%
  dplyr::mutate(orf_type = factor(orf_type, levels = c("uORF","novel","lncORF","dORF","ORF_annotated"))) %>%
  ggplot(aes(x = count, fill = orf_type)) +
  geom_histogram(stat = "count") +
  scale_x_continuous(breaks = 2:length(sample_ids)) +
  facet_wrap(~ orf_type, scales = "free_y") +
  theme_classic()
```

# FASTA

```{r tumoroid FASTA}
fasta_locs = grep("ORG",list.dirs("/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/analysis/ORFquant", recursive = F), value = T)
savedir = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/results"
interesting_orfs = read.delim(paste(savedir,"tumoroid","RMS_tumoroid_all_ORFs.txt",sep="/"), sep =";")
savefile = paste(savedir,"tumoroid","tumoroid_sequence_table.txt", sep = "/")

# code to extract all proteins from the file(s)
processed_ids <- character(0)
matching_seq <- character(0)
sequence_table <- data.frame(orf_id = character(0), sequence = character(0), width = integer(0), names = character(0), stringsAsFactors = FALSE)
filtered_interesting_orfs <- interesting_orfs %>%
  replace(is.na(x = .), "c") %>%
  dplyr::filter(!(class_code == "k"))

for (i in 1:nrow(filtered_interesting_orfs)) {
  id <- filtered_interesting_orfs$orf_id[i]
  source_file_number <- filtered_interesting_orfs$source[i]
  if (!(id %in% processed_ids)) {
    fasta_file <- list.files(grep(source_file_number,fasta_locs, value = T), pattern = "fasta", full.names = T)
    fasta <- Biostrings::readAAStringSet(fasta_file)
    matching_fasta <- fasta[grepl(id, names(fasta))]
    # Extract sequences, widths, and names
    matching_sequences <- as.character(matching_fasta)
    matching_widths <- width(matching_fasta)
    matching_names <- names(matching_fasta)
    # Add the extracted data to the sequence_table data frame
    if (length(matching_sequences) > 0) {
      new_rows <- data.frame(orf_id = id, sequence = matching_sequences, width = matching_widths, names = matching_names, stringsAsFactors = FALSE)
      sequence_table <- rbind(sequence_table, new_rows)
    }
    processed_ids <- c(processed_ids, id)
  }
}

tumoroid_fa <- Biostrings::AAStringSet(sequence_table[,c("sequence")])
names(tumoroid_fa) <- sequence_table[,c("names")]

Biostrings::writeXStringSet(tumoroid_fa, filepath = paste(savedir,"tumoroid","tumoroid_ORFs.fa", sep = "/"),
                            format = "fasta")
write.table(x = sequence_table, file = savefile,quote=F,sep="\t", row.names = F)

```

add missing files
```{r}
filtered_interesting_orfs <- interesting_orfs[!(interesting_orfs$orf_id %in% sequence_table$orf_id),]

# code to extract all proteins from the file(s)
processed_ids <- character(0)
matching_seq <- character(0)
sequence_table <- data.frame(orf_id = character(0), sequence = character(0), width = integer(0), names = character(0), stringsAsFactors = FALSE)

for (i in 1:nrow(filtered_interesting_orfs)) {
  id <- filtered_interesting_orfs$orf_id[i]
  source_file_number <- filtered_interesting_orfs$source[i]
  if (!(id %in% processed_ids)) {
    fasta_file <- list.files(grep(source_file_number,fasta_locs, value = T), pattern = "fasta", full.names = T)
    fasta <- Biostrings::readAAStringSet(fasta_file)
    matching_fasta <- fasta[grepl(id, names(fasta))]
    # Extract sequences, widths, and names
    matching_sequences <- as.character(matching_fasta)
    matching_widths <- width(matching_fasta)
    matching_names <- names(matching_fasta)
    # Add the extracted data to the sequence_table data frame
    if (length(matching_sequences) > 0) {
      new_rows <- data.frame(orf_id = id, sequence = matching_sequences, width = matching_widths, names = matching_names, stringsAsFactors = FALSE)
      sequence_table <- rbind(sequence_table, new_rows)
    }
    processed_ids <- c(processed_ids, id)
  }
}

tumoroid_fa <- Biostrings::AAStringSet(sequence_table[,c("sequence")])
names(tumoroid_fa) <- sequence_table[,c("names")]

Biostrings::writeXStringSet(tumoroid_fa, filepath = paste(savedir,"tumoroid","tumoroid_missed_ORFs.fa", sep = "/"),
                            format = "fasta")
write.table(x = sequence_table, file = paste(savedir,"tumoroid","tumoroid_missed_seqs.txt"),quote=F,sep="\t", row.names = F)
```


## Subset on interesting proteins

```{r}

prio_orf <- single_orf %>%
  replace(is.na(x = .), "c") %>%
  dplyr::filter(orf_category %in% c("uORF","dORF","novel","lncORF") &
                    !(class_code == "k"))

tum_seq_table <- read.delim(file = paste(savedir,"tumoroid/tumoroid_sequence_table.txt", sep = "/"), sep = " ") %>%
  dplyr::filter(width > 18 & 
                  width < 151 &
                  orf_id %in% prio_orf$orf_id)

sum(tum_seq_table$width)

tumoroid_fa <- Biostrings::AAStringSet(tum_seq_table[,c("sequence")])
names(tumoroid_fa) <- gsub("\\|.*","",tum_seq_table[,c("names")])

Biostrings::writeXStringSet(tumoroid_fa, filepath = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/03_prediction/data/tumoroid_sORF.fasta",
                            format = "fasta")

```
