---
title: "RMS patient ORFs"
author: "JD"
date: "2023-08-28"
output: html_document
---

TODO: annotated DF with protein sequence

```{r libraries}
library(dplyr)
library(rtracklayer)
library(GenomicRanges)
library(magrittr)
library(data.table)
```

```{r parameters}
savedir = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/results"
txome_gtf = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/01_rnaseq/analysis/rnaseq_pipeline/customannotation/RMS_full_novel_filtered_corrected.gtf"
txome_rannot = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/01_rnaseq/analysis/rnaseq_pipeline/customannotation/RMS_full/RMS_full_novel_filtered_corrected.gtf_Rannot"
orf_combined = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/results/orfs_combined.RData"
orf_overlap = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/results/orfs_overlap.RData"
orf_separate = "/hpc/pmc_vanheesch/projects/Jip/rms_analysis/02_riboseq/results/orfs_separate.RData"
```

This code block creates reference info for our little ORFs

```{r GTF based reference}
gtf_data <- data.table::fread(txome_gtf, skip = 5)
colnames(gtf_data) <- c("seqname", "source", "feature", "start", "end", 
                        "score", "strand", "frame", "attribute")
tid_gid <- gtf_data %>% 
    mutate(transcript_id = as.character(str_match(attribute, "ENST\\d+|MSTRG[.]\\d+[.]\\d+")),
           gene_id = as.character(str_match(attribute, "ENST\\d+|MSTRG[.]\\d+[.]\\d+"))) %>% 
  select(transcript_id, gene_id) %>% 
  filter(!is.na(transcript_id)) %>% distinct()

gtf_exon <- gtf_data %>% 
  mutate(transcript_id = as.character(str_match(attribute, "ENST\\d+|MSTRG[.]\\d+[.]\\d+")),
         gene_id = as.character(str_match(attribute, "ENST\\d+|MSTRG[.]\\d+[.]\\d+"))) %>% 
  filter(feature == "exon") %>% select(seqname, start, end, strand, 
                                       transcript_id, gene_id)

lnc_genes <- gtf_data %>% 
  mutate(gene_id = as.character(str_match(attribute, "ENST\\d+|MSTRG[.]\\d+[.]\\d+"))) %>% 
  filter(feature == "gene") %>% 
  mutate(biotype = as.character(
    str_match(attribute,"(?<=gene_biotype \").*(?=\")"))) %>% 
  filter(biotype == "lncRNA") %>% 
  select(gene_id, biotype)

#Load GTF rannot file from ORFquant
load(txome_rannot)
gtf_ref <- GTF_annotation$cds_txs %>% as.data.frame() %>% 
    group_by(group_name) %>% 
    summarise(gtf_sta = min(start), gtf_end = max(end)) %>% 
    dplyr::rename(transcript_id = group_name)
```

First, we load the data into R

```{r load data}

# sample IDs
sample_ids <- gsub("-.*","",basename(list.dirs("/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/riboseq_patientsamples/data/processed/trimgalore/", recursive = F)))

# TXOME GTF
tx_gtf <- rtracklayer::import.gff(txome_gtf)
tx_df <- as.data.frame(tx_gtf)

# ORF data
load(orf_separate)
load(orf_combined)
load(orf_overlap)

### ORF matrix
# Merge ORFs with 80% overlap, only keep the single parent as a reference
mat_80 <- orfquant_overlap[,grepl("0[.]8",colnames(orfquant_overlap))]
rownames(mat_80) <- names(orfquant_orfs_combined) 
mat_80 <- as.data.frame(mat_80) %>%
  dplyr::mutate(orf_id = names(orfquant_orfs_combined),
                count = rowSums(.[1:15]!=0))
colnames(mat_80) <- c(sample_ids,"parent","orf_id","count")

```

Use the following DEseq2 result tables from the RNA-seq to filter for potential
NBL-specific genes

```{r Check NBL-specificity}

nbl_canon <- read.delim("/hpc/pmc_vanheesch/projects/Jip/custom_transcriptomes/20221020_JD_quant_tumor_cohorts/analysis/NBL/results/NBL_canon_enriched_matrix.txt", sep = ";")

nbl_canon_pass <- nbl_canon %>% dplyr::filter(pass == T & selected == T)

nbl_novel <- read.delim("/hpc/pmc_vanheesch/projects/Jip/custom_transcriptomes/20221020_JD_quant_tumor_cohorts/analysis/NBL/results/NBL_novel_enriched_matrix.txt", sep = ";")

nbl_novel_pass <- nbl_novel %>% dplyr::filter(pass == T)

```

Stitch all the information together in a big table

```{r parse data into DF }

# Annotate the ORF IDs with gene-level data
orf_df <- data.frame(orf_id = names(orfquant_orfs_combined)) %>%
  dplyr::mutate(transcript_id = gsub("_.*","",names(orfquant_orfs_combined))) %>%
  dplyr::left_join(tx_df[tx_df$type == "transcript",
                         c("class_code","gene_name","gene_biotype","strand","transcript_id","gene_id")]) %>%
  dplyr::distinct()

# Create DF with ORF metadata
categories_all <- data.frame(mcols(orfquant_orfs_combined)) %>%
  dplyr::select(-gene_id,-transcript_id)

interesting_orfs <- orf_df %>%
  dplyr::left_join(categories_all, by = "orf_id") %>%
  # Only use novel ORFs outside existing CDS
  dplyr::filter(orf_type %in% c("dORF","novel","uORF") | 
                  gene_biotype %in% c("stringtie","lncRNA")) %>%
  dplyr::filter(
  # Find ORFs that occur in at least 2 samples
  orf_id %in% mat_80[which(mat_80$count > 1),"orf_id"]
  ) %>%
  
  # Append sample occurrence per ORF ID
  dplyr::left_join(mat_80[which(mat_80$parent == 0),c("orf_id","count")]) %>%
  
  # Annotate NBL-specificity
  dplyr::mutate(pass = ifelse(gene_id %in% c(nbl_canon_pass$gene_id,
                                             nbl_novel_pass$gene_id),T,F)) %>%
  dplyr::distinct()


```

```{r write table}

write.table(interesting_orfs, file = paste(savedir,"NBL_patient_novel_ORF.txt",sep="/"), quote = F, row.names = F, sep =";")


```


