---
title: "RMS merged combined ORF"
author: "JD"
date: "2024-02-15"
output: html_document
---


# Load libraries
```{r}
message("Loading required libraries ...")
suppressPackageStartupMessages({
  library(tidyverse)
  library(data.table)
  library(GenomicRanges)
  library(genomation)
  library(RColorBrewer)
  library(parallel)
})
```

# Define directories, files and colors
```{r}
# The following variables need to be defined:
gtf_file <- "/hpc/pmc_vanheesch/projects/jvandinter/rms_analysis/01_rnaseq/analysis/rnaseq_pipeline/customannotation/RMS_full_novel_filtered_corrected.sorted.gtf" # File: reference GTF
gtf_rannot_file <- paste("/hpc/pmc_vanheesch/projects/jvandinter/rms_analysis/01_rnaseq/analysis/rnaseq_pipeline/customannotation/","RMS_container","RMS_full_novel_filtered_corrected.gtf_Rannot", sep = "/") # File: Rannot file created by ORFquant
wd <- "/hpc/pmc_vanheesch/projects/jvandinter/rms_analysis/02_riboseq"
savedir <- paste(wd, "results", sep = "/")

orfquant_files <- "/hpc/pmc_vanheesch/projects/jvandinter/rms_analysis/02_riboseq/analysis/ORFquant/RMS_merged_psites/RMS_merged_psites_final_ORFquant_results" # File vector: *final_ORFquant_results ORFquant output
price_files <- "/hpc/pmc_vanheesch/projects/jvandinter/rms_analysis/02_riboseq/analysis/price/RMS_merged_price.orfs.bed" # File vector: *cit.bed PRICE output files

colors = c(brewer.pal(8, "Set1"))
```

# Define functions filtered
```{r functions}
prepare_orfs <- function(file_list, orf_method, ncores) {
  # Return a grangeslist with metadata columns representing the ORFs.
  #
  # Parameters:
  #   file_list - list, the input files from an ORF caller.
  #   orf_method - string, name of the ORF caller.
  #   ncores - int, number of cores to use.
  #
  # Returns:
  #   orfs - grangeslist, detected ORFs including the stop codon.
  if (orf_method == "PRICE") {
    orfs <- prepare_price(file_list, ncores)
    orfs <- lapply(orfs, function(x) {
      orf <- correct_price_geneid(x)
      m_cols <- mcols(orf)
      orf <- orf %>% as.data.frame() %>% mutate(group_name = name) %>% 
        makeGRangesListFromDataFrame(split.field = "group_name") 
      mcols(orf) <- m_cols
      return(orf)
    })
  } else if (orf_method == "ORFquant") {
    orfs <- prepare_orfquant(file_list)
    orfs <- lapply(orfs, function(orf) {
      orf <- add_stop_codon(orf, gtf_exon)
      return(orf)
    })
  } else if (orf_method == "Ribo-TISH") {
    orfs <- get_tish_orfs(file_list, gtf_exon, ncores)
  } else if (orf_method == "Ribotricer") {
    tricer_ref <- get_all_tricer_orfs(tricer_allorf_file)
    orfs <- prepare_ribotricer(file_list, tricer_ref, ncores)
    orfs <- mclapply(orfs, function(orf) {
      orf <- add_stop_codon(orf, gtf_exon)
      return(orf)
    }, mc.cores = ncores)
  }
  orfs <- mclapply(orfs, function(orf) {
    orf_cats <- orf %>% as.data.frame() %>% 
        mutate(transcript_id = as.character(str_match(group_name, "ENST\\d+|MSTRG[.]\\d+[.]\\d+"))) %>% 
        group_by(group_name, transcript_id, strand) %>% 
        summarise(orf_sta = min(start), orf_end = max(end), 
                  .groups = "keep") %>% annotate_categories(gtf_ref) %>% 
      ungroup() %>% select(group_name, orf_category)
    mcols(orf) <- mcols(orf) %>% as.data.frame() %>% 
      left_join(orf_cats, c("orf_id" = "group_name"))
    return(orf)
  }, mc.cores = ncores)
  return(orfs)
}

annotate_categories <- function(orf_seq, gtf_cds) {
  # Determine ORF categories based on transcript and GTF annotation
  #
  # Parameters:
  #   orf_seq - dataframe, ORFs with transcript ids and genomic start and end 
  #     coordinates
  #   gtf_cds - dataframe, gtf based reference with transcript ids and genomic 
  #     start and end coordinates
  #   ncores - int, number of cores to use.
  #
  # Returns:
  #   orf_cat_compare - dataframe, input dataframe with added orf_category 
  #     column
  orf_cat_compare <- orf_seq %>% left_join(gtf_cds, "transcript_id") %>% 
    mutate(orf_category = case_when(
      gtf_sta == orf_sta & gtf_end == orf_end ~ "ORF_annotated",
      gtf_sta > orf_sta & gtf_end == orf_end & strand == "+" ~ "N_extension",
      gtf_sta > orf_sta & gtf_end == orf_end & strand == "-" ~ "C_extension",
      gtf_sta < orf_sta & gtf_end == orf_end & strand == "+" ~ "N_truncation",
      gtf_sta < orf_sta & gtf_end == orf_end & strand == "-" ~ "C_truncation",
      gtf_end < orf_sta & gtf_end < orf_end & strand == "+" ~ "dORF",
      gtf_end < orf_sta & gtf_end < orf_end & strand == "-" ~ "uORF",
      gtf_sta < orf_sta & gtf_end < orf_end & strand == "+" ~ "overl_dORF",
      gtf_sta < orf_sta & gtf_end < orf_end & strand == "-" ~ "overl_uORF",
      gtf_sta > orf_sta & gtf_sta > orf_end & strand == "+" ~ "uORF",
      gtf_sta > orf_sta & gtf_sta > orf_end & strand == "-" ~ "dORF",
      gtf_sta > orf_sta & gtf_end > orf_end & strand == "+" ~ "overl_uORF",
      gtf_sta > orf_sta & gtf_end > orf_end & strand == "-" ~ "overl_dORF",
      gtf_sta < orf_sta & gtf_end > orf_end ~ "nested_ORF",
      gtf_sta > orf_sta & gtf_end < orf_end ~ "NC_extension",
      gtf_sta == orf_sta & gtf_end < orf_end & strand == "+" ~ "C_extension",
      gtf_sta == orf_sta & gtf_end < orf_end & strand == "-" ~ "N_extension",
      gtf_sta == orf_sta & gtf_end > orf_end & strand == "+" ~ "C_truncation",
      gtf_sta == orf_sta & gtf_end > orf_end & strand == "-" ~ "N_truncation",
      TRUE ~ "novel"
    )
  )
}

add_stop_codon <- function(orf_granges, gtf_exon) {
  # Extend all input ORFs to include the stop codon
  #
  # Parameters:
  #   orf_granges - GRangesList, ORFs to which a stop codon needs to be added
  #   gtf_exon - dataframe, gtf based reference annotation
  #
  # Returns:
  #   orf_shift - GRangesList, same as input, but with stopcodon included
  m_cols <- mcols(orf_granges)
  orf_df <- orf_granges %>% as.data.frame() %>% 
    mutate(transcript_id = as.character(str_match(group_name, "ENST\\d+|MSTRG[.]\\d+[.]\\d+"))) %>% 
    dplyr::rename(orf_start = start, orf_end = end, orf_strand = strand)
  orf_plus <- orf_df %>% filter(orf_strand == "+") %>% 
    inner_join(gtf_exon, "transcript_id", relationship = "many-to-many") %>% 
    filter(start <= orf_start, end >= orf_end) %>% 
    group_by(group_name) %>% 
    mutate(new_end = ifelse(orf_end == max(orf_end), orf_end + 3, orf_end),
           dif = ifelse(new_end > end, new_end - end, NA))
  orf_plus_extra <- orf_plus %>% filter(!is.na(dif)) %>% 
    left_join(gtf_exon, c("transcript_id", "strand"), 
              relationship = "many-to-many") %>% 
    filter(start.y > orf_start) %>% 
    group_by(group_name, dif, seqnames, strand) %>% 
    summarise(orf_start = min(start.y), .groups = "keep") %>% 
    mutate(new_end = orf_start + dif - 1) %>% ungroup() %>% select(-dif)
  orf_plus_shift <- orf_plus %>% 
    select(group_name, seqnames, strand, orf_start, new_end) %>% 
    rbind(orf_plus_extra) %>% dplyr::rename(start = orf_start, end = new_end)
  
  orf_minus <- orf_df %>% filter(orf_strand == "-") %>% 
    inner_join(gtf_exon, "transcript_id", relationship = "many-to-many") %>% 
    filter(start <= orf_start, end >= orf_end) %>% 
    group_by(group_name) %>% 
    mutate(new_start = ifelse(orf_start == min(orf_start), orf_start - 3, 
                              orf_start),
           dif = ifelse(new_start < start, start - new_start, NA))
  orf_minus_extra <- orf_minus %>% filter(!is.na(dif)) %>% 
    left_join(gtf_exon, c("transcript_id", "strand"), 
              relationship = "many-to-many") %>% 
    filter(end.y < orf_end) %>% 
    group_by(group_name, dif, seqnames, strand) %>% 
    summarise(orf_end = max(end.y), .groups = "keep") %>% 
    mutate(new_start = orf_end - dif + 1) %>% ungroup() %>% select(-dif)
  orf_minus_shift <- orf_minus %>% 
    select(group_name, seqnames, strand, new_start, orf_end) %>% 
    rbind(orf_minus_extra) %>% dplyr::rename(start = new_start, end = orf_end)
  orf_shift <- rbind(orf_plus_shift, orf_minus_shift) %>% distinct() %>% 
    makeGRangesListFromDataFrame(split.field = "group_name") 
  orf_shift <- orf_shift[m_cols$orf_id]
  mcols(orf_shift) <- m_cols
  return(orf_shift)
}

prepare_price <- function(files, ncores) {
  # Read PRICE files into GRangesLists
  #
  # Parameters:
  #   files - vector, contains the file paths to the PRICE .cit.bed files
  #
  # Returns:
  #   orfs_list - list, contains GRangesLists with the PRICE orfs
  orfs_list <- mclapply(files, function(x) readTranscriptFeatures(x)$exons, 
                        mc.cores = ncores)
  orfs_list <- lapply(orfs_list, function(x) split(x, x$name))
  for (i in 1:length(files)) {
    price_tsv_file <- str_replace(files[[i]], ".bed", ".tsv")
    price_tsv <- read.table(file = price_tsv_file, sep = '\t', header = TRUE)
    orfs_list[[i]] <- create_mcols(orfs_list[[i]], price_tsv)
  }
  return(orfs_list)
}

create_mcols <- function(price_orflist, price_table) {
  # Create the metadata columns for PRICE
  #
  # Parameters:
  #   price_orflist - GRangesList, ORFs detected by PRICE
  #   price_table - data frame, created from the orfs.tsv file created by PRICE
  #
  # Returns:
  #   price_orflist - GRangesList, with mcols added
  orf_mcols <- data.frame(names(price_orflist))
  orf_mcols <- orf_mcols %>% 
    separate(1,c("Gene_id", "Transcript_id", "Type", "Type_index", 
                 "StartCodon"), sep="__")
  orf_mcols$orf_id <- paste(orf_mcols$Transcript_id, orf_mcols$Type, 
                            orf_mcols$Type_index, sep = "_")
  orf_mcols <- left_join(orf_mcols, price_table[c(2, 7, 9:ncol(price_table))], 
                         c("orf_id"="Id"))
  orf_mcols <- dplyr::rename(orf_mcols, p_value = p.value, "p_sites" = Total, 
                             transcript_id = Transcript_id, gene_id = Gene_id)
  orf_mcols[c(7:ncol(orf_mcols))] <- sapply(orf_mcols[c(7:ncol(orf_mcols))], 
                                            as.numeric)
  orf_mcols$orf_id <- names(price_orflist)
  mcols(price_orflist) <- orf_mcols
  return(price_orflist)
}

correct_price_geneid <- function(orfs_price) {
  # Corrects PRICE gene ids for ORFs with mutliple gene ids assigned
  #
  # Parameters:
  #   orfs_price - GRangesLists, the detected ORFs by PRICE
  #
  # Returns:
  #   orfs_price - GRangesLists, same as input, but now each ORF has one gene id
  m_cols <- data.frame(mcols(orfs_price))
  gene_ids <- m_cols$gene_id
  transcript_ids <- m_cols$transcript_id
  m_cols$gene_id <- sapply(c(1:length(orfs_price)), 
    function(x) get_correct_geneid(transcript_ids[x], gene_ids[x]))
  mcols(orfs_price) <- m_cols
  return(orfs_price)
}

get_correct_geneid <- function(t_id, g_id) {
  # Retrieves correct gene id if there are multiple gene ids
  #
  # Parameters:
  #   t_id - string, the transcript id
  #   g_id - string, gene id determined by PRICE
  #   gid_tid_table - data frame, maps transcript ids to gene ids
  #
  # Returns:
  #   correct_gene_id - string, g_id if it consisted of a single gene id, the 
  #                     gene id found based on the transcript id if it consisted
  #                     of multiple gene ids
  split_id <- str_split(g_id, "_", simplify = T)
  if (length(split_id) == 1) {
    return(g_id)
  }
  correct_gene_id <- tid_gid$gene_id[which(tid_gid$transcript_id == t_id)]
  return(correct_gene_id)
}

prepare_orfquant <- function(files) {
  # Read ORFquant files into GRangesLists
  #
  # Parameters:
  #   files - vector, contains the file paths to the ORFquant RData files
  #
  # Returns:
  #   orfs_list - list, contains GRangesLists with the ORFquant orfs
  orfs_list <- vector(mode = "list", length = length(files))
  for (i in 1:length(files)) {
    load(files[i])
    orfs <- split(ORFquant_results$ORFs_gen, names(ORFquant_results$ORFs_gen))
    gene_ids <- data.frame(transcript_id = 
                             ORFquant_results$ORFs_tx$transcript_id, 
                           gene_id = ORFquant_results$ORFs_tx$gene_id, 
                           orf_type = ORFquant_results$ORFs_tx$ORF_category_Tx,
                           orf_id = names(ORFquant_results$ORFs_tx),
                           p_sites = ORFquant_results$ORFs_tx$P_sites_raw,
                           p_value = ORFquant_results$ORFs_tx$pval)
    orf_ids <- data.frame(orf_id = names(orfs))
    mcols(orfs) <- left_join(orf_ids, gene_ids, "orf_id")
    orfs_list[[i]] <- orfs
  }
  return(orfs_list)
}

get_combined_orfs <- function(orfs_list, files) {
  # Combine a list of GRangesLists into a single GRangeList
  #
  # Parameters:
  #   orfs_list - list, contains GRangesLists with the ORFs
  #   files - vector, contains the file paths to the input files used to create
  #                   orfs_list
  #
  # Returns:
  #   orfs_list_combined - GRangesLists, contains GRangesLists with the ORFquant 
  #                        orfs. Has an extra metacolumn source with the index
  #                        indicating from which sample the ORF was
  for (i in 1:length(orfs_list)) {
    orf_mcols <- mcols(orfs_list[[i]])
    orf_mcols$source <- rep(i, NROW(orfs_list[[i]]))
    mcols(orfs_list[[i]]) <- orf_mcols
  }
  orfs_list_unique <- lapply(orfs_list, remove_duplicates)
  orfs_list_combined <- do.call("c", orfs_list_unique)
  duplicate_hits <- findOverlaps(orfs_list_combined, orfs_list_combined, type = "equal")
  duplicate_count <- table(table(duplicate_hits@from))
  duplicate_count <- duplicate_count / as.integer(names(duplicate_count))
  sample_counts <- data.frame(sample=sapply(files, basename, USE.NAMES = FALSE),
                              all_orfs=sapply(orfs_list, NROW), 
                              unique_orfs=sapply(orfs_list_unique, NROW))
  print(sample_counts)
  print(paste0("All ORFs total: ", sum(sample_counts$all_orfs)))
  print(paste0("Unique ORFs total: ", sum(sample_counts$unique_orfs)))
  print("Counts of ORFs occuring in multiple samples:")
  print(duplicate_count)
  print(paste0("Sum of unique ORFs: ", sum(duplicate_count)))
  return(orfs_list_combined)
}

remove_duplicates <- function(g_range_list) {
  # Removes duplicates from the given GRangesList based on exact overlap
  #
  # Parameters:
  #   g_range_list - GRangesList
  #
  # Returns:
  #   g_range_unique - GRangesList, same as input, but without duplicates
  duplicate_hits <- findOverlaps(g_range_list, g_range_list, type = "equal")
  if (length(duplicate_hits) == length(g_range_list)) {
    return(g_range_list)
  }
  g_range_unique <- g_range_list[-duplicate_hits@to[
    duplicated(duplicate_hits@from)]]
  return(g_range_unique)
}

create_overlap_matrix <- function(com_orfs, thd, overlap_list, orf_lengths, 
                                  orf_sources, sample_num) {
  # Compute overlap between a set of ORFs for a give required overlap fraction
  #
  # Parameters:
  #   com_orfs - GRangesLists, the ORFs, including a source meta column
  #   thd - double, the required overlap fraction
  #   overlap_list - list, maps the indexes of overlapping ORFs
  #   orf_lengths - vector, the length of each ORF in com_orfs
  #   orf_sources - vector, the source index of each ORF in com_orfs
  #   sample_num - int, the total number of samples
  #
  # Returns:
  #   orfs_list_combined - GRangesLists, contains GRangesLists with the ORFquant 
  #                        orfs
  overlap_matrix <- matrix(0, ncol = sample_num + 1, nrow = NROW(com_orfs))
  for (i in 1:NROW(com_orfs)) {
    source_idx <- orf_sources[i]
    pos_parents <- overlap_list[[i]]
    overlap_matrix[i,source_idx] <- overlap_matrix[i,source_idx] + 1
    if (length(pos_parents) == 1) {
      next
    }
    child_width <- orf_lengths[i]
    len_parents <- sapply(pos_parents, function(x) child_width/orf_lengths[x])
    parent <- pos_parents[min(which(len_parents == 
                                      min(len_parents[len_parents >= thd])))]
    if (parent == i) {
      next
    }
    overlap_matrix[i,sample_num + 1] <- parent
    while (parent > 0) {
      overlap_matrix[parent, source_idx] = 
        overlap_matrix[parent, source_idx] + 1
      parent <- overlap_matrix[parent, sample_num + 1]
    }
  }
  parent_overlap_counts <- overlap_matrix[overlap_matrix[,sample_num + 1] == 
                                            0,1:sample_num]
  duplicate_counts <- table(rowSums(parent_overlap_counts > 0))
  print(paste0("Counts of ORFs occuring in multiple when requiring ",
               "minimum overlap of ",thd, ":"))
  print(duplicate_counts)
  print(paste0("Sum of unique ORFs: ", sum(duplicate_counts)))
  return(overlap_matrix)
}

get_partial_overlap <- function(com_orfs, thds) {
  # Computes full or partial overlap between a set of ORFs
  #
  # Parameters:
  #   com_orfs - GRangesLists, the ORFs, including a source meta column
  #   thds - vector, the required overlap fractions for which to compute the 
  #          overlap
  #
  # Returns:
  #   all_overlap - matrix, contains per threshold-source combination a column
  #                 named threshold_source and per threshold a column named
  #                 threshold_parent. Number of rows is equal to the length of
  #                 com_orfs. Per ORF and threshold it is indicated whether it
  #                 has an overlapping parent by an index in the parent column.
  #                 The threshold_source columns indicate which samples contain
  #                 an overlapping ORF.
  within_overlap <- findOverlaps(com_orfs, com_orfs, type = "within")
  minus_end_overlap <- findOverlaps(com_orfs[strand(com_orfs) == "-"], 
                                    com_orfs, type = "start")
  plus_end_overlap <- findOverlaps(com_orfs[strand(com_orfs) == "+"], 
                                   com_orfs, type = "end")
  within_overlap_list <- split(to(within_overlap), from(within_overlap))
  minus_end_overlap_list <- split(to(minus_end_overlap), 
                                  from(minus_end_overlap))
  plus_end_overlap_list <- split(to(plus_end_overlap), from(plus_end_overlap))
  unsorted_end_overlap_list <- c(minus_end_overlap_list, plus_end_overlap_list)
  end_overlap_list <- vector(mode = "list", 
                             length = length(unsorted_end_overlap_list))
  for (i in 1:length(unsorted_end_overlap_list)) {
    row_data <- unsorted_end_overlap_list[[i]]
    for (j in 1:length(row_data)) {
      end_overlap_list[[row_data[j]]] <- row_data
    }
  }
  overlap_list <- mapply(intersect, within_overlap_list, end_overlap_list)
  sample_num <- max(mcols(com_orfs)$source)
  all_overlap <- matrix(0, ncol = (sample_num+1) * length(thds), 
                        nrow = NROW(com_orfs))
  col_names_a <- rep(thds, each = sample_num+1)
  col_names_b <- rep.int(c(1:sample_num, "parent"), length(thds))
  colnames(all_overlap) <- paste(col_names_a, col_names_b, sep="_")
  orf_lengths <- sum(width(com_orfs))
  orf_sources <- mcols(com_orfs)$source
  for (i in 1:length(thds)) {
    overlap_matrix <- create_overlap_matrix(com_orfs, thds[i], overlap_list, 
                                            orf_lengths, orf_sources, sample_num)
    all_overlap[,((sample_num+1)*(i-1)+1):((sample_num+1)*i)] <- overlap_matrix
  }
  return(all_overlap)
}

get_overlap_counts <- function(m_overlap, thds, pred_source) {
  # Get count of shared ORFs between 1, 3 & 6 samples for each threshold 
  #
  # Parameters:
  #   m_overlap - matrix, created by get_partial_overlap function
  #   thds - vector, the required overlap fractions for which to compute the 
  #          overlap
  #   pred_source - string, name of the orfcaller
  #
  # Returns:
  #   graph_data - dataframe, contains for each combination of overlap fraction 
  #     and whether an ORF needs to be shared between 1, 3 & 6 samples the total
  #     number of unique orfs
  sample_num <- ncol(m_overlap)/length(thds) - 1
  graph_data <- matrix(0, ncol = 5, nrow = length(thds))
  colnames(graph_data) <- c("x", "Total", "Shared_3","Shared_6", "Source")
  graph_data <- as.data.frame(graph_data)
  graph_data[,"x"] <- thds
  for (i in 1:length(thds)) {
    overlap_cols <- sapply(1:sample_num, function(x) paste(thds[i], x, sep="_"))
    parent_col <- paste(thds[i], "parent", sep="_")
    overlap_counts <- m_overlap[m_overlap[,parent_col] == 0,overlap_cols]
    duplicate_counts <- table(rowSums(overlap_counts > 0))
    total_orf <- sum(duplicate_counts)
    common_orf <- sum(duplicate_counts[sample_num])
    half_orf <- sum(duplicate_counts[(sample_num-3):sample_num])
    graph_data[i,"Total"] <- total_orf
    graph_data[i, "Shared_6"] <- common_orf
    graph_data[i, "Shared_3"] <- half_orf
  }
  graph_data <- graph_data %>% pivot_longer(cols = c("Total", "Shared_6", "Shared_3")) %>% 
    mutate(Source = pred_source)
  return(graph_data)
}

filter_orfs <- function(m_overlap, m_orfs) {
  # Determine for each unique ORFs between how many samples it is shared
  # Parameters:
  #   m_overlap - matrix, created by get_partial_overlap function
  #   m_orfs - GRangesList, all ORFs from an ORF caller
  #
  # Returns:
  #   orfs_filtered - dataframe, rownames are ORF ids, the column indicates
  #     between how many samples the ORF was shared
  orfs_filtered <- m_overlap %>% 
    as.data.frame(row.names = names(m_orfs)) %>% 
    filter(`0.8_parent` == 0) %>% 
    dplyr::select(starts_with("0.8_") & !ends_with("parent")) %>% 
    apply(1, function(x) sum(x>0)) %>% 
    as.data.frame()
  names(orfs_filtered) <- "rowsum"
  return(orfs_filtered)
}

```

# Prepare GTF based references
```{r}
gtf_df <- rtracklayer::import(gtf_file) %>% 
  as.data.frame() %>%
  dplyr::select(gene_id,transcript_id,gene_name) %>%
  dplyr::distinct()

gtf_data <- fread(gtf_file, skip = 5)
colnames(gtf_data) <- c("seqname", "source", "feature", "start", "end", 
                        "score", "strand", "frame", "attribute")
tid_gid <- gtf_data %>% 
    mutate(transcript_id = as.character(str_match(attribute, "ENST\\d+|MSTRG[.]\\d+[.]\\d+")),
           gene_id = as.character(str_match(attribute, "ENSG\\d+|MSTRG[.]\\d+"))) %>% 
  select(transcript_id, gene_id) %>% 
  filter(!is.na(transcript_id)) %>% distinct()

gtf_exon <- gtf_data %>% 
  mutate(transcript_id = as.character(str_match(attribute, "ENST\\d+|MSTRG[.]\\d+[.]\\d+")),
         gene_id = as.character(str_match(attribute, "ENSG\\d+|MSTRG[.]\\d+"))) %>% 
  filter(feature == "exon") %>% select(seqname, start, end, strand, 
                                       transcript_id, gene_id)

lnc_genes <- gtf_data %>% 
  mutate(gene_id = as.character(str_match(attribute, "ENSG\\d+|MSTRG[.]\\d+"))) %>% 
  filter(feature == "gene") %>% 
  mutate(biotype = as.character(
    str_match(attribute,"(?<=gene_biotype \").*(?=\")"))) %>% 
  filter(biotype == "lncRNA") %>% 
  select(gene_id, biotype)

#Load GTF rannot file from ORFquant
load(gtf_rannot_file)
gtf_ref <- GTF_annotation$cds_txs %>% as.data.frame() %>% 
    group_by(group_name) %>% 
    summarise(gtf_sta = min(start), gtf_end = max(end)) %>% 
    dplyr::rename(transcript_id = group_name)

```

```{r}
thresholds <- c(0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1)

orf_methods <- c("ORFquant", "PRICE", "Ribo-TISH", "Ribotricer")

orfquant_orfs <- prepare_orfs(orfquant_files, orf_methods[1], 2)
price_orfs <- prepare_orfs(price_files, orf_methods[2], 2)

orfquant_orfs_combined <- get_combined_orfs(orfquant_orfs, orfquant_files)
price_orfs_combined <- get_combined_orfs(price_orfs, price_files)

orfquant_overlap <- get_partial_overlap(orfquant_orfs_combined, thresholds)
price_overlap <- get_partial_overlap(orfs_combined[[2]], thresholds)
```

Try adding PRICE & ORFquant results together and calculate combined ORFs

```{r}
# 
combined_grangeslist <- c(orfquant_orfs, price_orfs)
combined_orfs <- get_combined_orfs(combined_grangeslist, c(orfquant_files,price_files))
combined_orfs_overlap <- get_partial_overlap(combined_orfs, c(0.7,0.8,0.9,1.0)) %>%
  as.data.frame() %>%
  dplyr::mutate(orf_id = names(combined_orfs))

```

Some PRICE ORFs have an annotated non-canonical start codon, but actually have a ATG start very close to the annotated start.

1) Find ORFs with the same end coordinates on the same strand with the same gene ID
2) See if the start differs
```{r }
end_positions <- end(combined_orfs)
strands <- strand(combined_orfs)
# Step 1: Extract end positions and strands
end_positions <- lapply(combined_orfs, end)
strands <- lapply(combined_orfs, strand)

# Step 2: Split by end position and strand
split_grl <- split(grl, list(end_positions, strands))

```

Use ORF harmonisation script to combine ORFs from PRICE and ORFquant 
```{r keep ORFs}
orfs_keep <- filter_orfs(m_overlap = combined_orfs_overlap,
                      m_orfs = combined_orfs) %>%
  dplyr::mutate(orf_id = row.names(.))

mat <- combined_orfs_overlap[,grepl("^0.8_",colnames(combined_orfs_overlap))] %>%
  as.data.frame() %>%
  dplyr::rename(ORFquant = `0.8_1`,PRICE = `0.8_2`,parent = `0.8_parent`) %>%
  as.matrix()
rownames(mat) <- names(combined_orfs) 

mat <- as.data.frame(mat) %>%
  dplyr::mutate(orf_id = rownames(.)) %>%
  dplyr::filter(parent == "0") %>%
  dplyr::left_join(orfs_keep)
```

```{r}

```

Check effectiveness of harmonisation

```{r}

ppm_table <- read.delim("/hpc/pmc_vanheesch/projects/jvandinter/rms_analysis/02_riboseq/analysis/p_site_quantification/RMS_orfquant_merged_quant_psites.txt", sep = " ") %>%
  dplyr::mutate(psite_sum = rowSums(.),
                orf_id = rownames(.))
table(rowSums(ppm_table) == 0)

table(unique(orfquant_orfs$ORF_id) %in% mat$orf_id)
table(unique(price_orfs$name) %in% mat$orf_id)

subset_price <- price_orfs[which(price_orfs$name %in% mat$orf_id),]
subset_orfquant <- orfquant_orfs[which(orfquant_orfs$ORF_id %in% mat$orf_id),]

rtracklayer::export.gff(subset_orfquant, con = "/hpc/pmc_vanheesch/projects/jvandinter/rms_analysis/02_riboseq/results/combined/tmp_RMS_ORFquant.gtf")

rtracklayer::export.bed(subset_price, con = "/hpc/pmc_vanheesch/projects/jvandinter/rms_analysis/02_riboseq/results/combined/tmp_RMS_PRICE.bed")

```